---
title: "D19200 - Differential Expression"
author: "Kai"
date: "13th January 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)
library(ggplot2)
library(ggrepel)
library(MAST)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)

#create theme to not duplicate code
theme_jb <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5)
     ))
}

theme_jb_nogrid <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5),
       panel.grid = element_blank(),
       panel.border = element_blank(),
       axis.line = element_line()
     ))
}
```

# Load Seurat Object from preprocessing

```{r}
OUTPUT = "~/tscr/kai/projects/01_SCC_scRNA/results/samples/cohort/D19200/rp/"
load(paste0(OUTPUT,"RData/D19200-tumornofibmono.RData"))
```

# Optional: Determine Cell Cycle Phase

```{r}
#from Tirosh et al 2015 cell cycle genes, included in Seurat package
#The "data" slot is always picked, based on RNA assay
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
D19200.tumor <- CellCycleScoring(D19200.tumor,
                                 s.features = s.genes,
                                 g2m.features = g2m.genes, 
                                 assay="RNA")
```

# Differential Expression

As recommended by Seurat FAQ, we should use the RNA assay and NOT the pearson residuals of the SCT assay (inside of "scale.data" slot) https://github.com/satijalab/seurat/discussions/4032.

*Slot to use*
It is default and recommended to use the "data" slot of either RNA assay or SCT assay, since the counts are corrected for the sequencing depth. 
Mostly the log Foldchange is used in downstream analysis. The log Foldchange depends on the used slot! 
It is also to note, that the pct.1 and pct.2 values change if you use the RNA or SCT assay.

*Test to use*
Standard method is to use "wilcox". I use MAST since it is also recommended by Theis et al. (https://doi.org/10.15252/msb.20188746) and performs well in a general single cell DE review https://doi.org/10.1038/nmeth.4612, that was recommended in the Seurat tutorial. The test does not create as inflated p-values as with wilcox test.

*PCT option*
The general review also recommends a min.pct of 25%, but since this only excludes genes from testing, I would rather recommend the default of 0.1 to not miss too many information.

```{r}
# Set the Idents to the clustering or splitting of cells for which you want to perform DGE
Idents(D19200.tumor) <- NA
Idents(D19200.tumor) <- D19200.tumor$seurat_clusters_tumor_SCT_PCA

#Differential expression of every cluster against all other cluster
D19200.tumor.marker <- FindAllMarkers(D19200.tumor,
                                      assay = "RNA",
                                      slot="data", 
                                      test.use = "MAST", #alternative recommendation: MAST
                                      random.seed=1,
                                      logfc.threshold = 0, #shows the test results of all tested genes
                                      min.pct=0.1, #default; prior filtering of tested genes applied!
                                      latent.vars=NULL) # default; only for regression methods
```

# Volcano Plots 

```{r}
ggvolcanopct <- function(data,logFCthres) {
  logFCthres_low <- logFCthres[1]
  logFCthres_up <- logFCthres[2]
  ggplot(data = data,
         aes(x=avg_log2FC,
             y=pct.1-pct.2,
             col=avg_log2FC>logFCthres_up | avg_log2FC<logFCthres_low)) + 
    geom_point() + 
    geom_text_repel(aes(label=ifelse(avg_log2FC>logFCthres_up | avg_log2FC<logFCthres_low,as.character(gene),"")),
                    max.overlaps = 100) +
    scale_color_manual(values = c("TRUE"="red", "FALSE"="grey")) +
    scale_x_continuous(breaks=seq(-10,10,0.5)) +
    scale_y_continuous(breaks=seq(-1,1,0.1)) +
    coord_cartesian(ylim = c(-1,1), xlim=c(-max(abs(c(min(data$avg_log2FC),max(data$avg_log2FC))))-0.1,max(abs(c(min(data$avg_log2FC),max(data$avg_log2FC))))+0.1)) + #symmetric x-axis and y-axis with short buffer
    geom_hline(yintercept = 0, col="black", size=0.1) +
    geom_vline(xintercept = 0, col="black", size=0.1) +
    theme_jb() +
    theme(legend.position = "none", panel.grid.minor = element_blank())
}
#set thres function for selecting genes to mark in volcano plot
thres_func_plot = function(x) {
  c(sort(D19200.tumor.marker[D19200.tumor.marker$cluster==x,]$avg_log2FC, decreasing = F)[11], 
    sort(D19200.tumor.marker[D19200.tumor.marker$cluster==x,]$avg_log2FC, decreasing = T)[11])
}

#create volcano plots with pct difference on y-axis
D19200.tumor.marker.volpct <- lapply(levels(D19200.tumor.marker$cluster), function(x) ggvolcanopct(D19200.tumor.marker[D19200.tumor.marker$cluster==x,],thres_func_plot(x)) + ggtitle(x))
names(D19200.tumor.marker.volpct) <- levels(D19200.tumor.marker$cluster)

#save volcano plots
for (i in names(D19200.tumor.marker.volpct)) {
  ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-volcanopct-",i,".svg"),D19200.tumor.marker.volpct[[i]])
}

D19200.tumor.marker.volpct
```

# Gene Set Enrichment Analysis (GSEA)

```{r}
FCranks <- list()
for (cl in levels(D19200.tumor.marker$cluster)) {
  #rank DGE results by foldchange
  FCranks[[cl]] <- D19200.tumor.marker[D19200.tumor.marker$cluster==cl,]$avg_log2FC
  names(FCranks[[cl]]) <- D19200.tumor.marker[D19200.tumor.marker$cluster==cl,]$gene
  
  #convert -Inf to -1000
  FCranks[[cl]][is.infinite(FCranks[[cl]])] <- -1000
}
```

## Gene Ontology: Biological Processes

```{r}
library(fgsea)
#read in MsigDB pathway from downloads, in this case GO: BP
GOBP <- gmtPathways("~/tscr/kai/db/msigdb/c5.bp.v7.1.symbols.gmt")

GOBP.gsea <- list()
for (cl in levels(D19200.tumor.marker$cluster)) {
  #run fgsea
  GOBP.gsea[[cl]] <- fgsea(pathways=GOBP,
                           stats=FCranks[[cl]],
                           minSize=15,
                           maxSize=500,
                           nperm=10000)
}

#dataframe results
lapply(names(GOBP.gsea), function(x) cbind(x,GOBP.gsea[[x]][order(GOBP.gsea[[x]]$ES,decreasing = T),]))
```

## Gene Ontology: Hallmarks pathways

```{r}
library(fgsea)
HM <- gmtPathways("~/tscr/kai/db/msigdb/h.all.v7.1.symbols.gmt")

HM.gsea <- list()
for (cl in levels(D19200.tumor.marker$cluster)) {
  #run fgsea
  HM.gsea[[cl]] <- fgsea(pathways=HM,
                         stats=FCranks[[cl]],
                         minSize=15,
                         maxSize=500,
                         nperm=10000)
}

#dataframe results
lapply(names(HM.gsea), function(x) cbind(x,HM.gsea[[x]][order(HM.gsea[[x]]$ES,decreasing = T),]))
```

## Combined analysis of GSEA enrichment results

```{r}
#change names temporarily for interpretation
HM.gsea2 <- HM.gsea
GOBP.gsea2 <- GOBP.gsea

names(HM.gsea2) <- c("6 EMP2","3 EMP1","4 epi1","2 hypoxia","1 metabolism","7 cc","5 epi2")
names(GOBP.gsea2) <- c("6 EMP2","3 EMP1","4 epi1","2 hypoxia","1 metabolism","7 cc","5 epi2")

#save as supplementary CSVs
HM.gsea.sum <- as.data.frame(data.table::rbindlist(lapply(names(HM.gsea2), function(x) cbind(x,HM.gsea2[[x]][order(HM.gsea2[[x]]$ES,decreasing = T),]))))
HM.gsea.sum$leadingEdge <- unlist(lapply(HM.gsea.sum$leadingEdge,toString))
write.csv(x = HM.gsea.sum,file = paste0(OUTPUT,"D19200-tumornofibmono-diffexpr-GSEA-HM.csv"))

GOBP.gsea.sum <- as.data.frame(data.table::rbindlist(lapply(names(GOBP.gsea2), function(x) cbind(x,GOBP.gsea2[[x]][order(GOBP.gsea2[[x]]$ES,decreasing = T),]))))
GOBP.gsea.sum$leadingEdge <- unlist(lapply(GOBP.gsea.sum$leadingEdge,toString))
write.csv(x = GOBP.gsea.sum,file = paste0(OUTPUT,"D19200-tumornofibmono-diffexpr-GSEA-GOBP.csv"))
```

### Check some specific pathways in all clusters based on results

```{r}
#Enrichment plots for publication (based on GSEA results inclusion)

#EMT (SERPINE1, IGFBP3, TIMP3, DST, etc.)
plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`3`)
plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`1`)
plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`0`)
plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`5`)

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-2-hallmarkEMT.svg"),plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`3`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-3-hallmarkEMT.svg"),plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`1`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-6-hallmarkEMT.svg"),plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`0`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-7-hallmarkEMT.svg"),plotEnrichment(HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION ,FCranks$`5`),width=5,height=5)

#epithelial differentiation in epithelial 1 and epithelial 2
plotEnrichment(GOBP$GO_KERATINOCYTE_DIFFERENTIATION,FCranks$`2`)
plotEnrichment(GOBP$GO_KERATINOCYTE_DIFFERENTIATION,FCranks$`6`)

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-4-keratinocytedifferentiation.svg"),plotEnrichment(GOBP$GO_KERATINOCYTE_DIFFERENTIATION,FCranks$`2`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-5-keratinocytedifferentiation.svg"),plotEnrichment(GOBP$GO_KERATINOCYTE_DIFFERENTIATION,FCranks$`6`),width=5,height=5)

# hallmark hypoxia and glycolysis
plotEnrichment(HM$HALLMARK_GLYCOLYSIS,FCranks$`3`)
plotEnrichment(HM$HALLMARK_HYPOXIA,FCranks$`3`)

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-2-hallmarkglycolysis.svg"),plotEnrichment(HM$HALLMARK_GLYCOLYSIS,FCranks$`3`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-2-hallmarkhypoxia.svg"),plotEnrichment(HM$HALLMARK_HYPOXIA,FCranks$`3`),width=5,height=5)

#starvation and amino acid metabolism
plotEnrichment(GOBP$GO_RESPONSE_TO_STARVATION,FCranks$`4`)
plotEnrichment(GOBP$GO_ALPHA_AMINO_ACID_METABOLIC_PROCESS,FCranks$`4`)

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-1-responsestarvation.svg"),plotEnrichment(GOBP$GO_RESPONSE_TO_STARVATION,FCranks$`4`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-1-aminoacidmetabolism.svg"),plotEnrichment(GOBP$GO_ALPHA_AMINO_ACID_METABOLIC_PROCESS,FCranks$`4`),width=5,height=5)

#response to Interferon and Heat
plotEnrichment(GOBP$GO_RESPONSE_TO_TYPE_I_INTERFERON,FCranks$`1`)
plotEnrichment(GOBP$GO_REGULATION_OF_CELLULAR_RESPONSE_TO_HEAT,FCranks$`1`)

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-3-repsonsetypeIinterferon.svg"),plotEnrichment(GOBP$GO_RESPONSE_TO_TYPE_I_INTERFERON,FCranks$`1`),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-GSEAenrich-3-repsonseheat.svg"),plotEnrichment(GOBP$GO_REGULATION_OF_CELLULAR_RESPONSE_TO_HEAT,FCranks$`1`),width=5,height=5)
```

### get top 5 GSEA results per cluster

```{r}
#top 5 gene set based on ES of each cluster of results with p-value below 0.05 results
pvalthreshold <- 0.05
allenrichedpathways <- unique(unlist(lapply(GOBP.gsea, function(x){
  x[x$padj < pvalthreshold,][order(x[x$padj < pvalthreshold,]$ES,decreasing = T),]$pathway[1:5]})))
allenrichedpathways <- allenrichedpathways[!is.na(allenrichedpathways)]

allenrichedpathways
```

# Heatmap of differentially expressed genes

## Heatmaps with all genes above a certain logFC threshold

Load Pseudotime information
```{r}
#sorted by pseudotime; load in pseudotime object from other script
D19200.tumor.diffexpr <- D19200.tumor
load(paste0(OUTPUT,"RData/D19200-tumornofibmono-pseudotimeseurat.RData"))
ptcrv1 <- D19200.tumor$pca_SCT.crv1
ptcrv1 <- ptcrv1[colnames(D19200.tumor)] #same ordering as heatmap data
ptcrv2 <- D19200.tumor$pca_SCT.crv2
ptcrv2 <- ptcrv2[colnames(D19200.tumor)] #same ordering as heatmap data
D19200.tumor <- D19200.tumor.diffexpr
rm(D19200.tumor.diffexpr)
```

Formatting
```{r}
#logFC threshold
logfcthres.hm <- 0.5

#for colors
colorfunc <- function(n) {
  ifelse(n>9,
         colors <- rainbow(n = n),
         colors <- hue_pal()(n))
  colors
}
#Some data formatting for the heatmap
#format to gene-wise list of diffexpr results
D19200.tumor.marker.pergene <- lapply(unique(D19200.tumor.marker$gene), function(x) D19200.tumor.marker[D19200.tumor.marker$gene==x,])
names(D19200.tumor.marker.pergene) <- unique(D19200.tumor.marker$gene)

#select only entries in genes with logFC > logFC threshold
positivelogFCgenes <- lapply(D19200.tumor.marker.pergene, function(x) x[x$avg_log2FC>logfcthres.hm,])

#overview table of how many genes are in how many states differentially expressed
table(unlist(lapply(positivelogFCgenes,nrow)))
#discard all empty entries, i.e., genes that are not differentially expressed
positivelogFCgenes <- positivelogFCgenes[unlist(lapply(positivelogFCgenes,nrow))!=0]
#converge multiple entries and clusters
positivelogFCgenes.final <- lapply(positivelogFCgenes,function(x) {
  ifelse(nrow(x)>1,
         y <- data.frame(gene=unique(x$gene),
                         cluster=toString(x$cluster),
                         avg_log2FC=toString(x$avg_log2FC),
                         numberofstates = nrow(x)),
         y <- cbind(x[,c("gene","cluster","avg_log2FC")], numberofstates = nrow(x)))
  y
})
positivelogFCgenes.final <- as.data.frame(data.table::rbindlist(positivelogFCgenes.final))

#split up al duplicated genes again expect for  "0,5" and "2,6" since these have lots of overlapping genes
uebertrag <- positivelogFCgenes.final[positivelogFCgenes.final$numberofstates==1 | positivelogFCgenes.final$cluster %in% c("0, 5","2, 6"),]
uebertrag$cluster <- as.character(uebertrag$cluster)
tosplit <- positivelogFCgenes.final[positivelogFCgenes.final$numberofstates>1 & !positivelogFCgenes.final$cluster %in% c("0, 5","2, 6"),]

tosplit$cluster <- as.character(tosplit$cluster)

splitted <- as.data.frame(data.table::rbindlist(lapply(1:nrow(tosplit),function(x) {
  data.table::rbindlist(lapply(1:tosplit[x,]$numberofstates, function(y) {
    temp <- tosplit[x,] 
    temp$cluster <- strsplit(temp$cluster,split=", ")[[1]][y]
    temp$avg_log2FC <- strsplit(temp$avg_log2FC,split=", ")[[1]][y]
    temp$numberofstates <- 1
    temp
  }))
})))

positivelogFCgenes.final.split <- rbind(uebertrag,splitted)
positivelogFCgenes.final.split$cluster <- factor(positivelogFCgenes.final.split$cluster)

#for sorting heatmap by foldchange
positivelogFCgenes.final.split.sort <- positivelogFCgenes.final.split
positivelogFCgenes.final.split.sort[positivelogFCgenes.final.split.sort$numberofstates>1,]$avg_log2FC <- unlist(lapply(strsplit(positivelogFCgenes.final.split.sort[positivelogFCgenes.final.split.sort$numberofstates>1,]$avg_log2FC,split = ", "),function(x) x[1]))
positivelogFCgenes.final.split.sort$avg_log2FC <- as.numeric(positivelogFCgenes.final.split.sort$avg_log2FC)
```

```{r}
#for annotation of cluster, always name the top10 (top5?) genes by avglog2FC of this cluster
top10genes <- lapply(levels(positivelogFCgenes.final.split$cluster), function(clus) {
  #for those in > 1 cluster, get two results
  df <- positivelogFCgenes.final.split[positivelogFCgenes.final.split$cluster==clus,]
  if(max(df$numberofstates)>1) {
    fctemp <- strsplit(df$avg_log2FC,split = ", ")
    ordering <- lapply(1:length(fctemp[[1]]), function(x) order(as.numeric(unlist(lapply(fctemp, function(y) y[x]))),decreasing = T))
    top10genes <- lapply(1:length(fctemp[[1]]), function(x)  head(df[ordering[[x]],],n=10)$gene)
  } else {
    top10genes <- head(df[order(df$avg_log2FC,decreasing = T),],n=10)$gene
  }
  top10genes
})
  
names(top10genes) <- levels(positivelogFCgenes.final.split$cluster)
top10genes
```

### Heatmaps

```{r}
clustercolors <- colorfunc(length(unique(Idents(D19200.tumor))))
names(clustercolors) <- levels(Idents(D19200.tumor))

#bottom annotation
bottomanno.pub.crv1 <- HeatmapAnnotation(df = data.frame(cluster=Idents(D19200.tumor)[!is.na(ptcrv1)],
                                                         mito=D19200.tumor$percent.mito[!is.na(ptcrv1)],
                                                         Phase=D19200.tumor$Phase[!is.na(ptcrv1)]),
                                         col = list(cluster = clustercolors,
                                                    mito = circlize::colorRamp2(breaks=c(0,max(D19200.tumor$percent.mito)),colors = c("black","yellow")),
                                                    Phase = c(G1=brewer.pal(3,"Set1")[1],
                                                              G2M=brewer.pal(3,"Set1")[2],
                                                              S=brewer.pal(3,"Set1")[3])),
                                         which = "column", annotation_name_side = "left")

bottomanno.pub.crv2 <- HeatmapAnnotation(df = data.frame(cluster=Idents(D19200.tumor)[!is.na(ptcrv2)],
                                                         mito=D19200.tumor$percent.mito[!is.na(ptcrv2)],
                                                         Phase=D19200.tumor$Phase[!is.na(ptcrv2)]),
                                         col = list(cluster = clustercolors,
                                                    mito = circlize::colorRamp2(breaks=c(0,max(D19200.tumor$percent.mito)),colors = c("black","yellow")),
                                                    Phase = c(G1=brewer.pal(3,"Set1")[1],
                                                              G2M=brewer.pal(3,"Set1")[2],
                                                              S=brewer.pal(3,"Set1")[3])),
                                         which = "column", annotation_name_side = "left")

# new right annotation
annopathwaygenes <- list(amino_acid_metabolism=GOBP$GO_ALPHA_AMINO_ACID_METABOLIC_PROCESS,
                         response_starvation=GOBP$GO_RESPONSE_TO_STARVATION,
                         hallmark_hypoxia=HM$HALLMARK_HYPOXIA,
                         hallmark_glycolysis=HM$HALLMARK_GLYCOLYSIS,
                         hallmark_EMT=HM$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION,
                         response_typeI_interferon=GOBP$GO_RESPONSE_TO_TYPE_I_INTERFERON,
                         response_heat=GOBP$GO_REGULATION_OF_CELLULAR_RESPONSE_TO_HEAT,
                         keratinocyte_differentiation=GOBP$GO_KERATINOCYTE_DIFFERENTIATION,
                         G2Mphase=g2m.genes,
                         Sphase=s.genes)

rightanno.df.pub <- data.frame(data.table::rbindlist(lapply(names(annopathwaygenes), function(i) {data.frame(t(positivelogFCgenes.final.split$gene %in% annopathwaygenes[[i]]))})))
rownames(rightanno.df.pub) <- names(annopathwaygenes)
names(rightanno.df.pub) <- positivelogFCgenes.final.split$gene
rightanno.df.pub <- as.data.frame(t(rightanno.df.pub))
rightanno.df.pub.col <- lapply(names(rightanno.df.pub), function(x) c("TRUE"="grey25","FALSE"="white"))
names(rightanno.df.pub.col) <- names(rightanno.df.pub)

rightanno.pub <- HeatmapAnnotation(df=rightanno.df.pub,
                                   col = rightanno.df.pub.col,
                                   which="row",
                                   show_legend=F,
                                   annotation_name_gp = gpar(fontsize=11),
                                   annotation_name_side = "top",
                                   annotation_name_rot = 60)

clustersplit <- as.character(positivelogFCgenes.final.split$cluster)

#data and colors
expr.RNAscaledata <- D19200.tumor@assays$RNA@scale.data
expr.RNAscaledata.alldiffgenes.pub <- expr.RNAscaledata[as.character(positivelogFCgenes.final.split$gene),]

# set breaks for Heatmap colors
# Heatmap color breaks are set from 0 to the mean+2*sd of the absolute non-zero scaled expression onto the maximum value, therefore enables setting five colors in total
colorbreaks.RNAscaledata.pub <- c(-max(abs(expr.RNAscaledata.alldiffgenes.pub)),
                               -(mean(unlist(abs(expr.RNAscaledata.alldiffgenes.pub))[unlist(abs(expr.RNAscaledata.alldiffgenes.pub))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.alldiffgenes.pub))[unlist(abs(expr.RNAscaledata.alldiffgenes.pub))!=0])),
                               0,
                               mean(unlist(abs(expr.RNAscaledata.alldiffgenes.pub))[unlist(abs(expr.RNAscaledata.alldiffgenes.pub))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.alldiffgenes.pub))[unlist(abs(expr.RNAscaledata.alldiffgenes.pub))!=0]),
                               max(abs(expr.RNAscaledata.alldiffgenes.pub)))
#Heatmaps
diffexprhm.ptcrv1.pub.sort <- Heatmap(expr.RNAscaledata.alldiffgenes.pub[,!is.na(ptcrv1)],
                                     name="scale.data",
                                     col=circlize::colorRamp2(breaks=colorbreaks.RNAscaledata.pub, colors = c("red","purple", "black","yellow","white")),
                                     show_column_names = F,
                                     show_row_names = F,
                                     cluster_columns = F,
                                     cluster_column_slices = F,
                                     column_title_rot = 0,
                                     column_order = names(ptcrv1[!is.na(ptcrv1)][order(ptcrv1[!is.na(ptcrv1)])]),
                                     row_split = factor(clustersplit, levels=c("4","3","1","2","2, 6","6","0","0, 5","5")),
                                     cluster_rows = F,
                                     row_order = rownames(expr.RNAscaledata.alldiffgenes.pub)[order(positivelogFCgenes.final.split.sort$avg_log2FC,decreasing = T)],
                                     row_title_rot = 0,
                                     top_annotation = bottomanno.pub.crv1,
                                     right_annotation = rightanno.pub,
                                     #graphic options
                                     row_names_gp = gpar(fontsize=11),
                                     column_names_gp = gpar(fontsize=11),
                                     row_title_gp = gpar(fontsize=11),
                                     heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                                     border = T,
                                     use_raster = T,
                                     raster_device = "CairoPNG",
                                     raster_quality = 4,
                                     heatmap_height = unit(10,"in"),
                                     heatmap_width = unit(10,"in"))
#Save Heatmaps
pdf(paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-HM-RNAscaledata-logfc",as.character(logfcthres.hm),"-ptcrv1-pub-sort.pdf"), width=20, height=20)
diffexprhm.ptcrv1.pub.sort
dev.off()

diffexprhm.ptcrv2.pub.sort <- Heatmap(expr.RNAscaledata.alldiffgenes.pub[,!is.na(ptcrv2)],
                                     name="scale.data",
                                     col=circlize::colorRamp2(breaks=colorbreaks.RNAscaledata.pub, colors = c("red","purple", "black","yellow","white")),
                                     show_column_names = F,
                                     show_row_names = F,
                                     cluster_columns = F,
                                     cluster_column_slices = F,
                                     column_title_rot = 0,
                                     column_order = names(ptcrv2[!is.na(ptcrv2)][order(ptcrv2[!is.na(ptcrv2)])]),
                                     row_split = factor(clustersplit, levels=c("4","3","1","2","2, 6","6","0","0, 5","5")),
                                     cluster_rows = F,
                                     row_order = rownames(expr.RNAscaledata.alldiffgenes.pub)[order(positivelogFCgenes.final.split.sort$avg_log2FC,decreasing = T)],
                                     row_title_rot = 0,
                                     top_annotation = bottomanno.pub.crv2,
                                     right_annotation = rightanno.pub,
                                     #graphic options
                                     row_names_gp = gpar(fontsize=11),
                                     column_names_gp = gpar(fontsize=11),
                                     row_title_gp = gpar(fontsize=11),
                                     heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                                     border = T,
                                     use_raster = T,
                                     raster_device = "CairoPNG",
                                     raster_quality = 4,
                                     heatmap_height = unit(10,"in"),
                                     heatmap_width = unit(10,"in"))
#Save Heatmaps
pdf(paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-diffexpr-HM-RNAscaledata-logfc",as.character(logfcthres.hm),"-ptcrv2-pub-sort.pdf"), width=20, height=20)
diffexprhm.ptcrv2.pub.sort
dev.off()
```

# Specific differential expression

```{r}
#legend
#epi1: cluster 2
#epi2: cluster 6
#pEMT1: cluster 1
#pEMT2: Cluster 0

Idents(D19200.tumor) <- NA
Idents(D19200.tumor) <- D19200.tumor$seurat_clusters_tumor_SCT_PCA

pEMTvsepi <- FindMarkers(D19200.tumor,
                         ident.1 = c("2","6"),
                         ident.2= c("0","1"),
                         assay = "RNA",
                         slot="data", 
                         test.use = "MAST",
                         random.seed=1,
                         logfc.threshold = 0,
                         min.pct=0.1,
                         latent.vars=NULL)  

pEMTvsepi$gene <- rownames(pEMTvsepi)
```

## Volcano Plots 

```{r}
ggvolcanopctspecific <- function(data,logFCthres_col,logFCthres_text) {
  logFCthres_col_low <- logFCthres_col[1]
  logFCthres_col_up <- logFCthres_col[2]
  logFCthres_text_low <- logFCthres_text[1]
  logFCthres_text_up <- logFCthres_text[2]
  ggplot(data = data,
         aes(x=avg_log2FC,
             y=pct.1-pct.2,
             col=avg_log2FC>logFCthres_col_up | avg_log2FC<logFCthres_col_low)) + 
    geom_point() + 
    geom_text_repel(aes(label=ifelse(avg_log2FC>logFCthres_text_up | avg_log2FC<logFCthres_text_low,as.character(gene),"")),
                    max.overlaps = 100) +
    scale_color_manual(values = c("TRUE"="red", "FALSE"="grey")) +
    scale_x_continuous(breaks=seq(-signif(max(abs(pEMTvsepi$avg_log2FC))+1,digits = 1),
                                  signif(max(abs(pEMTvsepi$avg_log2FC))+1,digits = 1),
                                  signif((max(abs(data$avg_log2FC))+1)/20,digits=1))) +
    scale_y_continuous(breaks=seq(-1,1,0.1)) +
    coord_cartesian(ylim = c(-1,1), xlim=c(-max(abs(c(min(data$avg_log2FC),max(data$avg_log2FC))))-0.1,max(abs(c(min(data$avg_log2FC),max(data$avg_log2FC))))+0.1)) + #symmetric x-axis and y-axis with short buffer
    geom_hline(yintercept = 0, col="black", size=0.1) +
    geom_vline(xintercept = 0, col="black", size=0.1) +
    theme_jb() +
    theme(legend.position = "none", panel.grid.minor = element_blank())
}

pEMTvsepi.volpct <- ggvolcanopctspecific(pEMTvsepi,c(-1,1),c(sort(pEMTvsepi$avg_log2FC, decreasing = F)[11],
                                                                sort(pEMTvsepi$avg_log2FC, decreasing = T)[11]))

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-pEMTvsepi-SCT_PCA-volcanopct.svg"),pEMTvsepi.volpct + scale_x_continuous(breaks = seq(-20,20,0.5)), width = 5,height=5)

pEMTvsepi.volpct
```

# define gene signatures

p-EMT signature

```{r}
# decide upon average logFC of gene expression
#histogram
pEMTvsepi.log2FCplot <- ggplot() +
  geom_histogram(mapping = aes(x=pEMTvsepi$avg_log2FC, 
                 fill=pEMTvsepi$avg_log2FC>1 | pEMTvsepi$avg_log2FC<(-1)),
                 bins=100) +
  scale_x_continuous(breaks = seq(-10,10,0.5)) +
  scale_y_continuous(breaks = seq(0,600,150)) +
  scale_fill_manual(values = c("TRUE"="red", "FALSE"="black")) +
  coord_cartesian(xlim=c(-max(abs(c(min(pEMTvsepi$avg_log2FC),max(pEMTvsepi$avg_log2FC))))-0.1,max(abs(c(min(pEMTvsepi$avg_log2FC),max(pEMTvsepi$avg_log2FC))))+0.1)) +
  theme_jb() + 
  theme(legend.position = "none",panel.grid.minor = element_blank())

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-pEMTvsepi-SCT_PCA-log2FChist.svg"),pEMTvsepi.log2FCplot, width = 5,height=1.5)

#epithelial signature
epi.sig <- rownames(pEMTvsepi[pEMTvsepi$avg_log2FC>1,])
#p-EMT signature
pEMT.sig <- rownames(pEMTvsepi[pEMTvsepi$avg_log2FC<(-1),])
```

#Save

```{r}
#save signatures for GSVA
save(file = paste0(OUTPUT,"RData/D19200-tumornofibmono-diffexpr-sigs.RData"),pEMT.sig,epi.sig)
#save Seurat Object
save(file = paste0(OUTPUT,"RData/D19200-tumornofibmono-diffexpr-marker.RData"),D19200.tumor.marker)
#sessioninfo
writeLines(capture.output(sessionInfo()), paste0(OUTPUT,"RData/05-D19200-tumornofibmono-diffexpr-sessionInfo.txt"))
```
