---
title: "D19200 - analysis"
author: "Kai"
date: "13th January 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)
library(ggplot2)
library(ComplexHeatmap)
library(ggrepel)
library(GSVA)

theme_jb <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5),
       panel.grid.minor = element_blank()
     ))
}

theme_jb_nogrid <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5),
       panel.grid = element_blank(),
       panel.border = element_blank(),
       axis.line = element_line()
     ))
}

pseudotimeexpression <- function(gene=NULL,score=NULL,crv) {
  ifelse(!is.null(gene),
         expr <- GetAssayData(D19200.tumor,assay = "RNA",slot = "data")[gene,],
         expr <- D19200.tumor[[paste0(score,"1")]][,1])
  
  plotdata <- data.frame(cluster=D19200.tumor$seurat_clusters_tumor_SCT_PCA,
                         gene=expr,
                         pt=D19200.tumor[[paste0("pca_SCT.crv",crv)]][,1])
  ggplot(data=plotdata,
         aes(x=pt,y=gene, col=cluster)) +
    geom_point() +
    geom_smooth(method="gam", col="red", formula = y ~ s(x, bs = "cs")) +
    coord_cartesian(xlim = c(0,max(unlist(D19200.tumor[[c("pca_SCT.crv1","pca_SCT.crv2")]]),na.rm = T))) +
    scale_x_continuous(breaks = seq(0,200,10)) +
    labs(x=paste0("curve ",crv),y=ifelse(is.null(gene),score,gene)) +
    theme_jb()
}
```

# Load Seurat Object

```{r}
OUTPUT = "~/tscr/kai/projects/01_SCC_scRNA/results/samples/cohort/D19200/rp/"
load(paste0(OUTPUT,"RData/D19200-tumornofibmono-pseudotimeseurat.RData"))
```

# Change seurat clusters to new numbers

necessary for Violin Plots; numbers according to appaerance in trajectory

```{r}
D19200.tumor$trajectorycluster <- as.character(D19200.tumor$seurat_clusters_tumor_SCT_PCA)
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="0"] <- "6 temp"
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="1"] <- "3 temp"
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="2"] <- "4 temp"
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="3"] <- "2 temp"
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="4"] <- "1 temp"
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="5"] <- "7 temp"
D19200.tumor$trajectorycluster[D19200.tumor$trajectorycluster=="6"] <- "5 temp"
D19200.tumor$trajectorycluster <- sub(" temp","",D19200.tumor$trajectorycluster)
D19200.tumor$trajectorycluster <- factor(D19200.tumor$trajectorycluster)

clustercoltraj <- scales::hue_pal()(7)
names(clustercoltraj) <- c("6","3","4","2","1","7","5")
```

# Expression of own signatures

```{r}
# load own signatures
load(paste0(OUTPUT,"RData/D19200-tumornofibmono-diffexpr-sigs.RData"))
# add our own signature
ownsigs <- list()
ownsigs$OSCC_epithelial <- epi.sig
ownsigs$OSCC_EMT <- pEMT.sig

D19200.tumor <- AddModuleScore(D19200.tumor,
                               list(epi.sig),
                               name = "OSCC_epithelial",
                               assay = "RNA",
                               seed=1)
D19200.tumor <- AddModuleScore(D19200.tumor,
                               list(pEMT.sig),
                               name = "OSCC_EMT",
                               assay = "RNA",
                               seed=1)

ownsigs.umap <- FeaturePlot(D19200.tumor,
                            features = c("OSCC_epithelial1","OSCC_EMT1"),
                            reduction = "umap_SCT",
                            pt.size = 1,
                            combine = F)

ownsigs.vln <- VlnPlot(D19200.tumor,
                      features = c("OSCC_epithelial1","OSCC_EMT1"),
                      group.by = "trajectorycluster",
                      assay = "SCT",
                      slot = "data",
                      pt.size = 1,
                      combine = F)

names(ownsigs.umap) <- c("epi","EMT")
names(ownsigs.vln) <- c("epi","EMT")

#Visualize expression across pseudotime
epi.crv1.plot <- pseudotimeexpression(score="OSCC_epithelial",crv=1)
epi.crv2.plot <- pseudotimeexpression(score="OSCC_epithelial",crv=2)
EMT.crv1.plot <- pseudotimeexpression(score="OSCC_EMT",crv=1) 
EMT.crv2.plot <- pseudotimeexpression(score="OSCC_EMT",crv=2) 

lapply(names(ownsigs.umap), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-SCT-UMAP-",x,"-sig.svg"),ownsigs.umap[[x]] + theme_jb_nogrid(), width=5, height=5))
lapply(names(ownsigs.vln), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-SCTdata-VLNseuratclusters-",x,"-sig.svg"),ownsigs.vln[[x]] + theme_jb(), width=5, height=5))

lapply(names(ownsigs.umap), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-SCT-UMAP-",x,"-sig-noleg.svg"),ownsigs.umap[[x]] + theme_jb_nogrid() + theme(legend.position = "none") + ggtitle(NULL), width=5, height=5))
lapply(names(ownsigs.vln), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-SCTdata-VLNseuratclusters-",x,"-sig-noleg.svg"),ownsigs.vln[[x]] + theme_jb() + theme(legend.position = "none") + ggtitle(NULL), width=5, height=5))

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv1-episig.svg"),epi.crv1.plot,width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv2-episig.svg"),epi.crv2.plot,width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv1-EMTsig.svg"),EMT.crv1.plot,width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv2-EMTsig.svg"),EMT.crv2.plot,width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv1-episig-noleg.svg"),epi.crv1.plot + theme(legend.position = "none"),width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv2-episig-noleg.svg"),epi.crv2.plot + theme(legend.position = "none"),width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv1-EMTsig-noleg.svg"),EMT.crv1.plot + theme(legend.position = "none"),width = 2.5,height = 2.5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-ownsig-crv2-EMTsig-noleg.svg"),EMT.crv2.plot + theme(legend.position = "none"),width = 2.5,height = 2.5)

ownsigs.umap
ownsigs.vln
epi.crv1.plot
epi.crv2.plot
EMT.crv1.plot
EMT.crv2.plot
```

# Expression of examplified genes across pseudotime

```{r}
#differentially expressed genes
load(paste0(OUTPUT,"RData/D19200-tumornofibmono-diffexpr-marker.RData"))
#extract top 10 genes by logfc
D19200.tumor.marker.percluster <- lapply(unique(D19200.tumor.marker$cluster), function(x) D19200.tumor.marker[D19200.tumor.marker$cluster==x,])
names(D19200.tumor.marker.percluster) <- unique(D19200.tumor.marker$cluster)
topgenes <- lapply(names(D19200.tumor.marker.percluster), function(cl) {head(D19200.tumor.marker.percluster[[cl]][order(D19200.tumor.marker.percluster[[cl]]$avg_log2FC,decreasing = T),]$gene, n=10) })
# summarize in one dataframe
topgenes <- unlist(topgenes)

#manually defined important genes (e.g. EMT genes)
impgenes <- c("SFN","KRT14","KRT5","KRT17","HSPA1A","APOBEC3A","APOBEC3B","PSCA","ENO1","ENO2","ENO3","ALDOA","PGK1","PFKP","NDRG1","SERPINE1","LAMB3","TIMP3","VIM","CDH1","CDH2","TGFBI","EPCAM","CDH13","FN1")

#transcription factors
TFs <- c("ZEB1", "ZEB2", "TWIST1", "TWIST2","SNAI1","SNAI2","HIF1A")

allgenes <- unique(c(topgenes,impgenes,TFs))
  
#plot aacross pseudotime trajectory
ptexpr.crv1 <- lapply(allgenes, function(gene) pseudotimeexpression(gene=gene,crv=1))
names(ptexpr.crv1) <- paste0("crv1-",allgenes)
ptexpr.crv2 <- lapply(allgenes, function(gene) pseudotimeexpression(gene=gene,crv=2))
names(ptexpr.crv2) <- paste0("crv2-",allgenes)
ptexpr <- c(ptexpr.crv1,ptexpr.crv2)

lapply(names(ptexpr), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-",x,".svg"),ptexpr[[x]],width = 2.5,height = 2.5))
lapply(names(ptexpr), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-",x,"-noleg.svg"),ptexpr[[x]] + theme(legend.position = "none"),width = 2.5,height = 2.5))

#supplementary genes in larger size
impgenes_supp <- c("FN1","CDH2","MMP1","MMP13","IFITM1","EGR1","S100A7","S100A9","KRT16","KLK5","KLK6","KRT13","KRT15","THBS1","MT2A","ENO2","ENO3","ALDOA","PGK1","PFKP","EGLN3","ASNS")

ptexpr.crv1.l <- lapply(impgenes_supp, function(gene) pseudotimeexpression(gene=gene,crv=1))
names(ptexpr.crv1.l) <- paste0("crv1-",impgenes_supp)
ptexpr.crv2.l <- lapply(impgenes_supp, function(gene) pseudotimeexpression(gene=gene,crv=2))
names(ptexpr.crv2.l) <- paste0("crv2-",impgenes_supp)
ptexpr.l <- c(ptexpr.crv1.l,ptexpr.crv2.l)

lapply(names(ptexpr.l), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-",x,".svg"),ptexpr.l[[x]],width = 5,height = 5))
lapply(names(ptexpr.l), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-ptexpr-",x,"-noleg.svg"),ptexpr.l[[x]] + theme(legend.position = "none"),width = 5,height = 5))


#TFs as Violin Plot
TFs.vln <- VlnPlot(D19200.tumor,
                  features = TFs,
                  group.by = "trajectorycluster",
                  cols = clustercoltraj,
                  assay = "RNA",
                  slot = "data",
                  pt.size = 1,
                  combine = F)
TFs.vln <- lapply(TFs.vln, function(x) x + labs(x="cluster",y="expression") + theme_jb_nogrid())
names(TFs.vln) <- TFs

lapply(names(TFs.vln), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-VLNtraj-TF-",x,".svg"),TFs.vln[[x]],width = 4,height = 2.5))
lapply(names(TFs.vln), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-VLNtraj-TF-",x,"-noleg.svg"),TFs.vln[[x]] + theme(legend.position = "none") + ggtitle(NULL),width = 4,height = 2.5))

#percentage of cells expressing certain TF
expr.RNAdata <- D19200.tumor@assays$RNA@data
TF.perc <- unlist(lapply(TFs, function(TF) sum(expr.RNAdata[TF,]>0) / length(expr.RNAdata[TF,]>0)))
names(TF.perc) <- TFs

TF.perc.barplot <- ggplot(mapping = aes(x=names(TF.perc),y=TF.perc)) +
  geom_bar(stat="identity") +
  coord_cartesian(ylim=c(0,1)) +
  scale_y_continuous(breaks=seq(0,1,0.1)) +
  labs(x="transcription factor", y="% expr. cells") +
  theme_jb()
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-TF-perc.svg"),TF.perc.barplot,width = 4,height = 2.5)
```

# Analysis of EMTome signatures

## Load signatures

```{r}
library(fgsea)
#read in MsigDB pathway from downloads, in this case GO: BP
EMTome <- gmtPathways("~/tscr/kai/projects/01_SCC_scRNA/db/EMTome/EMTome_signatures.gmt")
#remove spaces
EMTome <- lapply(EMTome,function(x) gsub(" ","",x))
#rename Rokavic et al since it is there two times
names(EMTome)[which(names(EMTome)=="Rokavec_et_al.2017")[2]] <- "Rokavec_et_al.2017_2"
#rename remsik
names(EMTome)[grep("Rem",names(EMTome))] <- "Remsik_et_al.2018"

#Add signatures to EMTome database
## Puram et al signatures
pEMT.genes <- read.table("~/tscr/kai/projects/01_SCC_scRNA/db/Purametal/PARTIAL_EMT.txt")
pEMT.genes <- pEMT.genes$V1
EMTome$Puram_et_al_partialEMT <- pEMT.genes
epi1.genes <- read.table("~/tscr/kai/projects/01_SCC_scRNA/db/Purametal/EPITHELIAL_DIFFERENTIATION_1.txt")
epi1.genes <- epi1.genes$V1
EMTome$Puram_et_al_epithelial1 <- epi1.genes
epi2.genes <- read.table("~/tscr/kai/projects/01_SCC_scRNA/db/Purametal/EPITHELIAL_DIFFERENTIATION_2.txt")
epi2.genes <- epi2.genes$V1
EMTome$Puram_et_al_epithelial2 <- epi2.genes

# Kinker  et al. signatures
SuppTable4_tenRHPs <- read.csv("~/tscr/kai/projects/01_SCC_scRNA/db/kinker_tirosh_scRNAseqcelllinepaper/SuppTable4_tenRHPs.csv")
kinker_et_al <- lapply(1:ncol(SuppTable4_tenRHPs), function(x) as.character(SuppTable4_tenRHPs[,x])[as.character(SuppTable4_tenRHPs[,x])!=""])
names(kinker_et_al) <- paste0("Kinker_et_al_", names(SuppTable4_tenRHPs))

EMTome$Kinker_et_al_EMT.I <- kinker_et_al$Kinker_et_al_EMT.I
EMTome$Kinker_et_al_EMT.II <- kinker_et_al$Kinker_et_al_EMT.II
EMTome$Kinker_et_al_EMT.III <- kinker_et_al$Kinker_et_al_EMT.III
EMTome$Kinker_et_al_Epithelial.Senescence <- kinker_et_al$Kinker_et_al_Epithelial.Senescence
```

## Quality Control of signatures

```{r}
expr.RNAdata <- as.data.frame(GetAssayData(D19200.tumor, assay = "RNA", slot = "data"))

allsigs <- c(EMTome,ownsigs) #all signatures

allsigs.allgenes <- unique(unlist(allsigs)) # all genes in signatures
tumor.allgenes <- rownames(expr.RNAdata) #all genes in scRNAseq
tumor.exprgenes <- rownames(expr.RNAdata[rowSums(expr.RNAdata)!=0,]) # all expressed genes in scRNAseq

#genes in gene expression matrix (GEM)
genesinGEM.perc <- unlist(lapply(allsigs, function(x) sum(x %in% tumor.allgenes) / length(x %in% tumor.allgenes))) 
genesinGEM.abs <- unlist(lapply(allsigs, function(x) sum(x %in% tumor.allgenes)))
genesinGEM.names <- unlist(lapply(allsigs, function(x) toString(x[x %in% tumor.allgenes])))

#genes not expressed
genesnotexpr.perc <- unlist(lapply(allsigs, function(x) sum(x %in% tumor.exprgenes) / length(x %in% tumor.exprgenes)))
genesnotexpr.abs <- unlist(lapply(allsigs, function(x) sum(x %in% tumor.exprgenes)))
genesnotexpr.names <- unlist(lapply(allsigs, function(x) toString(x[x %in% tumor.exprgenes])))

allsigs.anno <- data.frame(signature=names(genesnotexpr.perc),
                          inGEM_perc_rank=rank(-genesinGEM.perc, ties.method = "first"),
                          inGEM_perc=genesinGEM.perc,
                          inGEM_abs_rank=rank(-genesinGEM.abs, ties.method = "first"),
                          inGEM_abs=genesinGEM.abs,
                          inGEM_abs_log10=log10(genesinGEM.abs),
                          inGEM_genes=genesinGEM.names,
                          EXPR_perc_rank=rank(-genesnotexpr.perc, ties.method = "first"),
                          EXPR_perc=genesnotexpr.perc,
                          EXPR_abs_rank=rank(-genesnotexpr.abs, ties.method = "first"),
                          EXPR_abs=genesnotexpr.abs,
                          EXPR_abs_log10=log10(genesnotexpr.abs),
                          EXPR_genes=genesnotexpr.names,
                          size=unlist(lapply(allsigs, length)),
                          size_log10=log10(unlist(lapply(allsigs, length))),
                          size_rank=rank(-unlist(lapply(allsigs, length)),ties.method = "first")
                          )
rownames(allsigs.anno) <- names(allsigs)

#Visualize size
library(ggrepel)
ggplot(data = allsigs.anno,
       mapping = aes(x=size_rank,y=size)) +
  geom_point() +
  theme_bw()

#Visualize no of genes in GEM
ggplot(data = allsigs.anno,
       mapping = aes(x=inGEM_abs_rank,y=inGEM_abs)) +
  geom_point() +
  theme_jb()

#Visualize percentage of genes in GEM
ggplot(data = allsigs.anno,
       mapping = aes(x=inGEM_perc_rank,y=inGEM_perc)) +
  geom_point() +
  geom_text_repel(aes(label=ifelse(inGEM_perc<0.9,as.character(signature),""))) + 
  coord_cartesian(ylim = c(0,1)) +
  scale_y_continuous(breaks=seq(0,1,0.05)) +
  theme_jb()

#Visualizae percentage of expressed genes
ggplot(data = allsigs.anno,
       mapping = aes(x=EXPR_perc_rank,y=EXPR_perc)) +
  geom_point() +
  geom_text_repel(aes(label=ifelse(EXPR_perc<0.8,as.character(signature),""))) + 
  coord_cartesian(ylim = c(0,1)) +
  scale_y_continuous(breaks=seq(0,1,0.05)) +
  theme_jb()
```

## Overlap if signatures and GSVA

```{r}
library(GSVA)
library(RColorBrewer)
#extended EMTome by own signatures
EMTome.ext <- c(EMTome,epi=list(ownsigs$OSCC_epithelial),EMT=list(ownsigs$OSCC_EMT))

#Calculate % of overlap between all signatures
signatureoverlap <- lapply(EMTome.ext,function(y) data.frame(t(unlist(lapply(EMTome.ext,function(x) length(intersect(y,x)) / length(y))))))
signatureoverlap <- as.data.frame(data.table::rbindlist(signatureoverlap))
rownames(signatureoverlap) <- names(signatureoverlap)

#GSVA
EMT.gsva <- gsva(expr= as.matrix(expr.RNAdata), 
                      gset.idx.list = c(EMTome,ownsigs),
                      method="gsva", #default
                      kcdf="Gaussian", #default
                      abs.ranking=F, #default
                      min.sz=1, #default
                      max.sz=Inf, #default
                      mx.diff=T #default
)

# Spearman correlation 
EMT.gsva.cor <- cor(t(EMT.gsva),method = "spearman")

#only with EMTome
EMT.gsva.emtome <- EMT.gsva[!rownames(EMT.gsva) %in% c("OSCC_epithelial","OSCC_EMT"),]
EMT.gsva.cor.emtome <- EMT.gsva.cor[!rownames(EMT.gsva.cor) %in% c("OSCC_epithelial","OSCC_EMT"),!colnames(EMT.gsva.cor) %in% c("OSCC_epithelial","OSCC_EMT")]
```

### Heatmaps

```{r}
# Correlation Heatmap only of EMTome signatures
p1 <- Heatmap(EMT.gsva.cor.emtome,
              name="correlation1",
              column_title = "EMTome signatures",
              column_title_side = "bottom",
              col=circlize::colorRamp2(breaks=c(-1,-0.5,0,0.5,1), colors = c("darkblue", "blue","white","red","darkred")),
              show_column_names = F,
              show_row_names = F,
              cluster_columns = T,
              clustering_method_columns = "ward.D2",
              clustering_distance_columns = "spearman",
              cluster_rows = T,
              clustering_method_rows =  "ward.D2",
              clustering_distance_rows = "spearman",
              #graphic options
              row_names_gp = gpar(fontsize=6),
              column_names_gp = gpar(fontsize=6),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(3.5,"in"),
              height = unit(3.5,"in")
)

#GSVA Scores across pseudotime
#Annotation
D19200.tumor <- CellCycleScoring(D19200.tumor,
                                 s.features = cc.genes$s.genes,
                                 g2m.features = cc.genes$g2m.genes,
                                 assay="RNA")
bottomanno.crv1.df <- data.frame(cluster=D19200.tumor$trajectorycluster,Phase=D19200.tumor$Phase)
rownames(bottomanno.crv1.df) <- names(D19200.tumor$trajectorycluster)
bottomanno.crv1.df <- bottomanno.crv1.df[names(sort(D19200.tumor$pca_SCT.crv1[!is.na(D19200.tumor$pca_SCT.crv1)])),]
bottomanno.crv1.pub <- HeatmapAnnotation(df = bottomanno.crv1.df,
                                col = list(cluster = clustercoltraj,
                                           Phase = c(G1=brewer.pal(3,"Set1")[1],G2M=brewer.pal(3,"Set1")[2],S=brewer.pal(3,"Set1")[3])),
                                which = "column", annotation_name_side = "left",annotation_name_gp = gpar(fontsize=11))
#Heatmap
p0 <- Heatmap(EMT.gsva.emtome[,names(sort(D19200.tumor$pca_SCT.crv1[!is.na(D19200.tumor$pca_SCT.crv1)]))],
              name="GSVA Score",
              column_title = "cells",
              column_title_side = "bottom",
              row_title = "EMTome signatures",
              row_title_side = "left",
              col=circlize::colorRamp2(breaks=c(-1,-0.5,0,0.5,1), colors = c("darkblue", "blue","white","red","darkred")),
              show_row_names = F,
              show_column_names = F,
              cluster_rows = F,
              cluster_columns = F,
              column_order = names(sort(D19200.tumor$pca_SCT.crv1[!is.na(D19200.tumor$pca_SCT.crv1)])),
              row_order = row_order(p1),
              bottom_annotation = bottomanno.crv1.pub,
              #graphic options
              row_title_rot = 90,
              row_names_gp = gpar(fontsize=11),
              column_names_gp = gpar(fontsize=9),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(3.5,"in"),
              height = unit(3.5,"in")
)

#Correlation plot for own signatures
allsigs.anno.EMTome <- allsigs.anno[rownames(EMT.gsva.emtome),]
rightanno.pub <- HeatmapAnnotation(df=allsigs.anno.EMTome[,c("EXPR_perc","size_log10")],
                                col = list(EXPR_perc=circlize::colorRamp2(breaks=c(0,1),colors = c("white","purple")),
                                           size_log10=circlize::colorRamp2(breaks=c(0,max(allsigs.anno.EMTome$size_log10)),colors = c("white","darkgreen"))),
                                which="row", annotation_name_side = "bottom",annotation_name_gp = gpar(fontsize=11))

p2 <- Heatmap(t(EMT.gsva.cor[c("OSCC_epithelial","OSCC_EMT"),!colnames(EMT.gsva.cor) %in% c("OSCC_epithelial","OSCC_EMT")]),
              name="correlation",
              col=circlize::colorRamp2(breaks=c(-1,-0.5,0,0.5,1), colors = c("darkblue", "blue","white","red","darkred")),
              show_column_names = T,
              show_row_names = F,
              cluster_columns = F,
              cluster_rows = F,
              row_order = row_order(p1),
              right_annotation = rightanno.pub,
              #graphic options
              column_title_rot = 90,
              row_names_gp = gpar(fontsize=6),
              column_names_gp = gpar(fontsize=11),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(0.5,"in"),
              height = unit(3.5,"in")
)

#all combined
pdf(paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-EMTgsva.pdf"), width=20, height=20)
p0 + p1 + p2
dev.off()

#overlap within EMTome
p3 <- Heatmap(signatureoverlap[!rownames(signatureoverlap) %in% c("epi","EMT"),!colnames(signatureoverlap) %in% c("epi","EMT")],
              name="%ov",
              col=circlize::colorRamp2(breaks=c(0,0.5,1), colors = c("white", "red","darkred")),
              show_column_names = F,
              show_row_names = F,
              cluster_columns = F,
              cluster_rows = F,
              column_order = row_order(p1),
              #graphic options
              column_title_rot = 90,
              row_names_gp = gpar(fontsize=6),
              column_names_gp = gpar(fontsize=6),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(3.5,"in"),
              height = unit(3.5,"in")
)

#overlap between own signatures and EMTome
p4 <- Heatmap(t(signatureoverlap[c("epi","EMT"),!colnames(signatureoverlap) %in% c("epi","EMT")]),
              name="%ov",
              col=circlize::colorRamp2(breaks=c(0,0.5,1), colors = c("white", "red","darkred")),
              show_column_names = T,
              show_row_names = F,
              cluster_columns = F,
              cluster_rows = F,
              #graphic options
              column_title_rot = 90,
              row_names_gp = gpar(fontsize=6),
              column_names_gp = gpar(fontsize=11),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(0.5,"in"),
              height = unit(3.5,"in")
)

pdf(paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-EMTgsva-ov.pdf"), width=20, height=20)
p0 + p3 + p4 + p1 + p2
dev.off()

#save Correlation and Overlap Heatmap with row/column names
p1.n <- Heatmap(EMT.gsva.cor.emtome,
              name="correlation1",
              column_title = "EMTome signatures",
              column_title_side = "bottom",
              col=circlize::colorRamp2(breaks=c(-1,-0.5,0,0.5,1), colors = c("darkblue", "blue","white","red","darkred")),
              show_column_names = T,
              show_row_names = T,
              cluster_columns = T,
              clustering_method_columns = "ward.D2",
              clustering_distance_columns = "spearman",
              cluster_rows = T,
              clustering_method_rows =  "ward.D2",
              clustering_distance_rows = "spearman",
              #graphic options
              row_names_gp = gpar(fontsize=6),
              column_names_gp = gpar(fontsize=6),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(7,"in"),
              height = unit(7,"in")
)

p3.n <- Heatmap(signatureoverlap[!rownames(signatureoverlap) %in% c("epi","EMT"),!colnames(signatureoverlap) %in% c("epi","EMT")],
              name="%ov",
              col=circlize::colorRamp2(breaks=c(0,0.5,1), colors = c("white", "red","darkred")),
              show_column_names = T,
              show_row_names = T,
              cluster_columns = F,
              cluster_rows = F,
              column_order = row_order(p1),
              row_order =  row_order(p1),
              #graphic options
              column_title_rot = 90,
              row_names_gp = gpar(fontsize=6),
              column_names_gp = gpar(fontsize=6),
              row_title_gp = gpar(fontsize=11),
              heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
              border = T,
              use_raster = T,
              raster_device = "CairoPNG",
              raster_quality = 4,
              width = unit(7,"in"),
              height = unit(7,"in")
)

pdf(paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-EMTgsva-cor.pdf"), width=20, height=20)
p1.n
dev.off()

pdf(paste0(OUTPUT,"figs-D19200/D19200-tumornofibmono-RNAdata-EMTome-ov.pdf"), width=20, height=20)
p3.n
dev.off()
```

# Save

```{r}
#save image
save(file = paste0(OUTPUT,"RData/D19200-tumornofibmono-analysis.RData"),D19200.tumor,EMT.gsva)
#save SessionInfo
writeLines(capture.output(sessionInfo()), paste0(OUTPUT,"RData/07-D19200-tumornofibmono-analysis-sessionInfo.txt"))
```
