---
title: "D19200 - tumor cc"
author: "Kai"
date: "13th January 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)
library(ggplot2)

theme_jb <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5)
     ))
}

theme_jb_nogrid <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5),
       panel.grid = element_blank(),
       panel.border = element_blank(),
       axis.line = element_line()
     ))
}
```

# Load Seurat Object from preprocessing

```{r}
OUTPUT = "~/tscr/kai/projects/01_SCC_scRNA/results/samples/cohort/D19200/rp/"
load(paste0(OUTPUT,"RData/D19200-tumornofibmono.RData"))
```

# Determine Cell Cycle Phase

```{r}
#from Tirosh et al 2015 cell cycle genes, included in Seurat package
#The "data" slot is always picked, based on RNA assay
s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes
D19200.tumor <- CellCycleScoring(D19200.tumor,
                                s.features = s.genes,
                                g2m.features = g2m.genes, 
                                assay="RNA")
```

# Normalization

## Normalization of RNA assay

```{r SCC_normalization}
#RNA normalization without regressing, necessary for example for differential gene expression not based on SCT.
DefaultAssay(D19200.tumor) <- "RNA"
#log-normalization for slot "data"
D19200.tumor <- NormalizeData(D19200.tumor, 
                              normalization.method = "LogNormalize", #default
                              scale.factor = 10000, #default, scale factor 10000 is recommendation from seurat tutorial
                              assay = "RNA",
                              margin = 1 # default; normalizes across features
) 

#Find Variable Features
D19200.tumor <- FindVariableFeatures(D19200.tumor, 
                                     assay = "RNA",
                                     selection.method = "vst", #default
                                     nfeatures = 2000 #default; only 2000 , 3000 for SCT promises since to "restore" some more VariableFeatures
)

#scaling for slot "scale.data"
allgenes <- rownames(D19200.tumor)
D19200.tumor <- ScaleData(D19200.tumor, 
                          features = allgenes, 
                          do.scale = T,
                          do.center = T,
                          scale.max = 10,
                          assay = "RNA",
                          vars.to.regress = c("S.Score","G2M.Score"))
```

## Normalization with SCTransform

More information: https://doi.org/10.1186/s13059-019-1874-1

```{r SCC_normalization}
D19200.tumor <- SCTransform(D19200.tumor, 
                            assay="RNA", #default
                            new.assay.name = "SCT", #default; overwrites old SCT
                            do.correct.umi = T, #default change counts slot to corrected counts in new assay
                            variable.features.n = 3000, #default set variable features
                            vars.to.regress = c("S.Score","G2M.Score"), #default optional variables to regress out
                            do.scale = F, #default scale pearson residuals in scale.data slot to have unit variance
                            do.center = T, #default center pearson residuals in scale.data to have mean zero, needed for PCA
                            return.only.var.genes = T, #default scale.data.matrices output assay only contains variable genes
                            seed.use = 1448145)
```

# Dimension reduction

## PCA

always based on SCT assay

```{r}
D19200.tumor <- RunPCA(D19200.tumor, 
                       npcs = 50,  #number of PCs to use
                       assay = "SCT",
                       rev.pca = F, # default, Run cell x gene matrix
                       weight.by.var = T, #default, Weight cell embedding by variance of each PC!)
                       approx = T, #if TRUE, uses irlba instead of prcomp, which is a single value decomposition to approximate PCs, saving computation time (by not calculating all PCs; important for big datasets) and being similar/same to PCs
                       features = VariableFeatures(D19200.tumor,assay = "SCT"), #default
                       reduction.name = "pca_SCT",
                       verbose = F)
```

```{r}
#calculate the % variance explained
PCAvar.SCT <- D19200.tumor@reductions$pca_SCT@stdev^2 / sum(matrixStats::rowVars(as.matrix(D19200.tumor@assays$SCT@scale.data)[VariableFeatures(D19200.tumor,assay = "SCT"),]))

PCAplot.SCT <- DimPlot(D19200.tumor, reduction = "pca_SCT", dims = c(1,2), group.by = "orig.ident") + 
  labs(x=paste0("principal component 1 (",signif(PCAvar.SCT[1]*100,digits = 2),"% variance)"),
       y=paste0("principal component 2 (",signif(PCAvar.SCT[2]*100,digits = 2),"% variance)")) +
  theme_jb()

PCAheatmap.SCT <- DimHeatmap(D19200.tumor, 
                             dims = 1:20, 
                             balanced = TRUE, 
                             fast=F, 
                             reduction = "pca_SCT",
                             assays = "SCT",
                             slot = "scale.data",
                             combine = F) 
PCAheatmap.SCT <- lapply(PCAheatmap.SCT, function(x) x + theme_jb() + theme(axis.text.x = element_blank(), axis.line.x = element_blank()))

PCAelbowplot.SCT <- ggplot(data=data.frame(var=PCAvar.SCT,
                                           PC=1:length(PCAvar.SCT))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=var)) +
  labs(x="principal component", y="% of variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

# cumulative Variance Plot
PCAsumplot.SCT <- ggplot(data=data.frame(cumvar=cumsum(PCAvar.SCT),PC=1:length(PCAvar.SCT))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=cumvar)) +
  labs(x="principal component", y="% of cumulative variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA.svg"),PCAplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-noleg.svg"),PCAplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
lapply(1:length(PCAheatmap.SCT), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-heatmap-PC",x,".svg"),PCAheatmap.SCT[[x]], height=5, width=5))
lapply(1:length(PCAheatmap.SCT), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-heatmap-PC",x,"-noleg.svg"),PCAheatmap.SCT[[x]] + theme(legend.position = "none"), height=5, width=5))
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-elbow.svg"),PCAelbowplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-elbow-noleg.svg"),PCAelbowplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-variancesum.svg"),PCAsumplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-variancesum-noleg.svg"),PCAsumplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)

PCAplot.SCT
PCAheatmap.SCT
PCAelbowplot.SCT
PCAsumplot.SCT
```

## Graph-based clustering

here we perform two different clusterings, one based on the PCA of all cells, one of the tumor.

- seurat_clusters_all_SCT_PCA:      "old" seurat clusters before subsetting (from preprocessing script)
- seurat_clusters_tumor_SCT_PCA:    New seurat clusters from subsetting performed on PCA on SCT assay

```{r FindClusters}
numberofPCs = 20

#based on PCA
D19200.tumor <- FindNeighbors(D19200.tumor, 
                              dims = 1:numberofPCs, 
                              reduction = "pca_SCT",
                              assay = "SCT",
                              force.recalc = T
)
D19200.tumor <- FindClusters(D19200.tumor, 
                                    resolution = 0.8, #default 
                                    random.seed=100)

D19200.tumor$cctraj <- D19200.tumor$seurat_clusters
D19200.tumor$seurat_clusters <- NULL 

#rename clusters according to non-regressed cluster characteristics (this is absed on information from differential expression and pseudotime analysis more downstream)
D19200.tumor$cctraj <- as.character(D19200.tumor$cctraj)
D19200.tumor$cctraj[D19200.tumor$cctraj=="0"] <- "6 temp"
D19200.tumor$cctraj[D19200.tumor$cctraj=="1"] <- "4 temp"
D19200.tumor$cctraj[D19200.tumor$cctraj=="2"] <- "7 temp"
D19200.tumor$cctraj[D19200.tumor$cctraj=="3"] <- "3 temp"
D19200.tumor$cctraj[D19200.tumor$cctraj=="4"] <- "2 temp"
D19200.tumor$cctraj[D19200.tumor$cctraj=="5"] <- "1 temp"
D19200.tumor$cctraj[D19200.tumor$cctraj=="6"] <- "5 temp"
D19200.tumor$cctraj <- sub(" temp","",D19200.tumor$cctraj)
D19200.tumor$cctraj <- factor(D19200.tumor$cctraj)

clustercol <- scales::hue_pal()(7)
names(clustercol) <- c("6","3","4","2","1","7","5")

#save new PCA with colored clusters 
PCAplot.SCT.cctraj <- DimPlot(D19200.tumor,
                            dims=1:2, 
                            reduction="pca_SCT", 
                            pt.size = 1, 
                            group.by = "cctraj") + 
         labs(x=paste0("principal component 1 (",signif(PCAvar.SCT[1]*100,digits = 2),"% of variance)"),
              y=paste0("principal component 2 (",signif(PCAvar.SCT[2]*100,digits = 2),"% of variance)")) +
  scale_color_manual(values = clustercol) +
  theme_jb_nogrid()
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-cctraj.svg"),PCAplot.SCT.cctraj, width = 5, height = 5)
PCAplot.SCT.cctraj

PCAplot.SCT.phase <- DimPlot(D19200.tumor,
                            dims=1:2, 
                            reduction="pca_SCT", 
                            pt.size = 1, 
                            group.by = "Phase") + 
         labs(x=paste0("principal component 1 (",signif(PCAvar.SCT[1]*100,digits = 2),"% of variance)"),
              y=paste0("principal component 2 (",signif(PCAvar.SCT[2]*100,digits = 2),"% of variance)")) +
  theme_jb_nogrid()
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-PCA-Phase.svg"),PCAplot.SCT.phase, width = 5, height = 5)
PCAplot.SCT.phase
```

## UMAP

```{r UMAP, fig.width=10, fig.height=8}
D19200.tumor <- RunUMAP(D19200.tumor,
                        dims = 1:numberofPCs,
                        assay = "SCT",
                        umap.method = "uwot", # Seurat default
                        graph=NULL, #default
                        reduction="pca_SCT",
                        reduction.name = "umap_SCT"
)

UMAP.SCT.cctraj <- DimPlot(D19200.tumor, 
                           reduction = "umap_SCT", 
                           pt.size = 1, 
                           label = T, 
                           group.by = "cctraj") + 
  scale_x_continuous(breaks = seq(-10,10,1)) +
  scale_y_continuous(breaks = seq(-10,10,0.5)) +
  scale_color_manual(values = clustercol) +
  theme_jb_nogrid()

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-UMAP-cctraj.svg"),UMAP.SCT.cctraj, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-UMAP-cctraj-noleg.svg"),UMAP.SCT.cctraj + theme(legend.position = "none") + ggtitle(""), width = 5, height = 5)

UMAP.SCT.phase <- DimPlot(D19200.tumor, 
                          reduction = "umap_SCT", 
                          pt.size = 1, 
                          label = F, 
                          group.by = "Phase") + 
  scale_x_continuous(breaks = seq(-10,10,1)) +
  scale_y_continuous(breaks = seq(-10,10,0.5)) +
  theme_jb_nogrid()

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-UMAP-phase.svg"),UMAP.SCT.phase, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-UMAP-phase-noleg.svg"),UMAP.SCT.phase + theme(legend.position = "none") + ggtitle(""), width = 5, height = 5)

UMAP.SCT.cctraj
UMAP.SCT.phase
```

Evaluate the amount of proliferating cells in certain subpopulations

```{r}
#Split the UMAP in three different regions for three main populations:
D19200.tumor$epinonepi <- as.character(D19200.tumor$orig.ident)
D19200.tumor$epinonepi[rownames(D19200.tumor@reductions$umap_SCT@cell.embeddings[D19200.tumor@reductions$umap_SCT@cell.embeddings[,1] < 2 & D19200.tumor@reductions$umap_SCT@cell.embeddings[,2] < 0.5,])] <- "pEMP"
D19200.tumor$epinonepi[rownames(D19200.tumor@reductions$umap_SCT@cell.embeddings[D19200.tumor@reductions$umap_SCT@cell.embeddings[,1] < 2 & D19200.tumor@reductions$umap_SCT@cell.embeddings[,2] >= 0.5,])] <- "metabolism"
D19200.tumor$epinonepi[rownames(D19200.tumor@reductions$umap_SCT@cell.embeddings[D19200.tumor@reductions$umap_SCT@cell.embeddings[,1] >= 2,])] <- "epi"

UMAP.SCT.epinonepi <- DimPlot(D19200.tumor, 
                          reduction = "umap_SCT", 
                          pt.size = 1, 
                          label = F, 
                          group.by = "epinonepi") + 
  scale_x_continuous(breaks = seq(-10,10,1)) +
  scale_y_continuous(breaks = seq(-10,10,0.5)) +
  theme_jb_nogrid()

UMAP.SCT.epinonepi

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-UMAP-epinonepi.svg"),UMAP.SCT.epinonepi, width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-UMAP-epinonepi-noleg.svg"),UMAP.SCT.epinonepi + theme(legend.position = "none") + ggtitle(""), width=5,height=5)

# Create barplot with fraction of proliferating cells
phaseplotdata <- data.frame(Phase=names(table(D19200.tumor$Phase[D19200.tumor$epinonepi=="epi"])),
                                      as.numeric(table(D19200.tumor$Phase[D19200.tumor$epinonepi=="epi"])),
                                      as.numeric(table(D19200.tumor$Phase[D19200.tumor$epinonepi=="metabolism"])),
                                      as.numeric(table(D19200.tumor$Phase[D19200.tumor$epinonepi=="pEMP"])))
names(phaseplotdata) <- c("Phase","epi. diff.","metabolism","partial EMP")
umapclus.phase <- ggplot(data=reshape2::melt(phaseplotdata)) + 
  geom_bar(aes(x=variable,y=value,fill=Phase),
           stat = "identity",
           position = "fill") +
  labs(x="",y="fraction of cells") +
  theme_jb() + 
  theme(axis.text.x = element_text(angle=30,hjust=1,vjust=1))

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-umapclus-phase.svg"),umapclus.phase, width=2.5,height=3)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-cc-tumornofibmono-SCT-umapclus-phase-noleg.svg"),umapclus.phase + theme(legend.position = "none") + ggtitle(""), width=2.5,height=3)

UMAP.SCT.epinonepi
umapclus.phase
```

## Save 

```{r}
#Single Seurat object
save(file = paste0(OUTPUT,"RData/D19200-cc-tumornofibmono.RData"),D19200.tumor)
#SessionInfo
writeLines(capture.output(sessionInfo()), paste0(OUTPUT,"RData/09-D19200-cc-tumornofibmono-sessionInfo.txt"))
```
