---
title: "D19200 - fibroendo"
author: "Kai"
date: "13th January 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)
library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
#create theme to not duplicate code
theme_jb <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5)
     ))
}

theme_jb_nogrid <- function(base_size=11) {
  (theme_bw() +
     theme(
       axis.title = element_text(size=11),
       axis.text = element_text(size=11),
       legend.text = element_text(size=11),
       #optional rotation of x-axis labels
       axis.text.x = element_text(angle=90, hjust=1,vjust=.5),
       panel.grid = element_blank(),
       panel.border = element_blank(),
       axis.line = element_line()
     ))
}
```

# Load Seurat Object from preprocessing

```{r}
OUTPUT = "~/tscr/kai/projects/01_SCC_scRNA/results/samples/cohort/D19200/rp/"
load(paste0(OUTPUT,"RData/D19200-fibro.RData"))
load(paste0(OUTPUT,"RData/D19200-filt.RData"))
```

# Subset Cells

Filter tumor and monocytes and add endothelial cells

```{r}
fibrocells <- names(D19200.f$seurat_clusters_fibro_SCT_PCA[!D19200.f$seurat_clusters_fibro_SCT_PCA %in% c("6","7")])
endocells <- names(D19200.filt$CellType[D19200.filt$CellType=="Endothelial Cells"])

D19200.ef <- subset(D19200.filt,cells = c(fibrocells,endocells))
```

# Normalization

After subsetting of cells we have to re-run the preprocessing steps for Normalization, dimensionality reduction, etc. 
This is needed, since the gene expression matrix is now reduced and enables more accurate capturing of the heterogeneity in the cells.
Important: The slots, reductions and SCT assay saved in D19200.ef are still from the non-subsetted object, therefore we will override these in the next steps.

## Normalization of RNA assay

```{r SCC_normalization}
#RNA normalization without regressing, necessary for example for differential gene expression not based on SCT.
DefaultAssay(D19200.ef) <- "RNA"
#log-normalization for slot "data"
D19200.ef <- NormalizeData(D19200.ef, 
                           normalization.method = "LogNormalize", #default
                           scale.factor = 10000, #default, scale factor 10000 is recommendation from seurat tutorial
                           assay = "RNA",
                           margin = 1 # default; normalizes across features
) 

#Find Variable Features
D19200.ef <- FindVariableFeatures(D19200.ef, 
                                  assay = "RNA",
                                  selection.method = "vst", #default
                                  nfeatures = 2000 #default; only 2000 , 3000 for SCT promises since to "restore" some more VariableFeatures
)

#scaling for slot "scale.data"
allgenes <- rownames(D19200.ef)
D19200.ef <- ScaleData(D19200.ef, 
                       features = allgenes, 
                       do.scale = T,
                       do.center = T,
                       scale.max = 10,
                       assay = "RNA")
```

## Normalization with SCTransform

More information: https://doi.org/10.1186/s13059-019-1874-1

```{r SCC_normalization}
D19200.ef <- SCTransform(D19200.ef, 
                         assay="RNA", #default
                         new.assay.name = "SCT", #default; overwrites old SCT
                         do.correct.umi = T, #default change counts slot to corrected counts in new assay
                         variable.features.n = 3000, #default set variable features
                         vars.to.regress = NULL, #default optional variables to regress out
                         do.scale = F, #default scale pearson residuals in scale.data slot to have unit variance
                         do.center = T, #default center pearson residuals in scale.data to have mean zero, needed for PCA
                         return.only.var.genes = T, #default scale.data.matrices output assay only contains variable genes
                         seed.use = 1448145)
```

# Dimension reduction

## PCA

always based on SCT assay

```{r}
D19200.ef <- RunPCA(D19200.ef, 
                    npcs = 50,  #number of PCs to use
                    assay = "SCT",
                    rev.pca = F, # default, Run cell x gene matrix
                    weight.by.var = T, #default, Weight cell embedding by variance of each PC!)
                    approx = T, #if TRUE, uses irlba instead of prcomp, which is a single value decomposition to approximate PCs, saving computation time (by not calculating all PCs; important for big datasets) and being similar/same to PCs
                    features = VariableFeatures(D19200.ef,assay = "SCT"), #default
                    reduction.name = "pca_SCT",
                    verbose = F)
```

```{r}
PCAvar.SCT <- D19200.ef@reductions$pca_SCT@stdev^2 / sum(matrixStats::rowVars(as.matrix(D19200.ef@assays$SCT@scale.data)[VariableFeatures(D19200.ef,assay = "SCT"),]))

PCAplot.SCT <- DimPlot(D19200.ef, reduction = "pca_SCT", dims = c(1,2), group.by = "orig.ident") + 
  labs(x=paste0("principal component 1 (",signif(PCAvar.SCT[1]*100,digits = 2),"% variance)"),
       y=paste0("principal component 2 (",signif(PCAvar.SCT[2]*100,digits = 2),"% variance)")) +
  theme_jb()

PCAheatmap.SCT <- DimHeatmap(D19200.ef, 
                             dims = 1:20, 
                             balanced = TRUE, 
                             fast=F, 
                             reduction = "pca_SCT",
                             assays = "SCT",
                             slot = "scale.data",combine = F) 
PCAheatmap.SCT <- lapply(PCAheatmap.SCT, function(x) x + theme_jb() + theme(axis.text.x = element_blank(), axis.line.x = element_blank()))

PCAelbowplot.SCT <- ggplot(data=data.frame(var=PCAvar.SCT,
                                           PC=1:length(PCAvar.SCT))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=var)) +
  labs(x="principal component", y="% variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

# cumulative Variance Plot
PCAsumplot.SCT <- ggplot(data=data.frame(cumvar=cumsum(PCAvar.SCT),PC=1:length(PCAvar.SCT))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=cumvar)) +
  labs(x="principal component", y="% cumulative variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()


ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA.svg"),PCAplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-noleg.svg"),PCAplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
lapply(1:length(PCAheatmap.SCT), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-heatmap-PC",x,".svg"),PCAheatmap.SCT[[x]], height=5, width=5))
lapply(1:length(PCAheatmap.SCT), function(x) ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-heatmap-PC",x,"-noleg.svg"),PCAheatmap.SCT[[x]] + theme(legend.position = "none"), height=5, width=5))
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-elbow.svg"),PCAelbowplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-elbow-noleg.svg"),PCAelbowplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-variancesum.svg"),PCAsumplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-variancesum-noleg.svg"),PCAsumplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)

PCAplot.SCT
PCAheatmap.SCT
PCAelbowplot.SCT
PCAsumplot.SCT
```

## Graph-based clustering

here we perform two different clusterings, one based on the PCA of all cells, one of the fibroblasts

- seurat_clusters_all_SCT_PCA:      "old" seurat clusters before subsetting (from preprocessing script)
- seurat_clusters_fibro_SCT_PCA:    New seurat clusters from subsetting performed on PCA on SCT assay

```{r FindClusters}
numberofPCs = 20

#based on PCA
D19200.ef <- FindNeighbors(D19200.ef, 
                           dims = 1:numberofPCs, 
                           reduction = "pca_SCT",
                           assay = "SCT")
D19200.ef <- FindClusters(D19200.ef, 
                          resolution = 1.2,
                          random.seed=100)

D19200.ef$seurat_clusters_fibro_SCT_PCA <- D19200.ef$seurat_clusters
D19200.ef$seurat_clusters <- NULL 

#save new PCA with colored clusters 
PCAplot.SCT.clus <- DimPlot(D19200.ef,
                            dims=1:2, 
                            reduction="pca_SCT", 
                            pt.size = 1, 
                            group.by = "seurat_clusters_fibro_SCT_PCA") + 
  labs(x=paste0("principal component 1 (",signif(PCAvar.SCT[1]*100,digits = 2),"% of variance)"),
       y=paste0("principal component 2 (",signif(PCAvar.SCT[2]*100,digits = 2),"% of variance)")) +
  theme_jb_nogrid()
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-clus.svg"),PCAplot.SCT.clus, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-PCA-clus.svg"),PCAplot.SCT.clus + theme(legend.position = "none") + labs(title = NULL) + ggtitle(NULL), width = 5, height = 5)
PCAplot.SCT.clus
```

## UMAP

```{r UMAP, fig.width=10, fig.height=8}
D19200.ef <- RunUMAP(D19200.ef,
                     dims = 1:numberofPCs,
                     assay = "SCT",
                     umap.method = "uwot", # Seurat default
                     graph=NULL, #default
                     reduction="pca_SCT",
                     reduction.name = "umap_SCT"
)

UMAP.SCT.clusPCA <- DimPlot(D19200.ef, 
                            reduction = "umap_SCT", 
                            pt.size = 1, 
                            label = T, 
                            group.by = "seurat_clusters_fibro_SCT_PCA") + 
  theme_jb_nogrid()

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-clusPCA.svg"),UMAP.SCT.clusPCA, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-clusPCA.svg"),UMAP.SCT.clusPCA + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
UMAP.SCT.clusPCA
```

# Characterization of fibroblast clusters

Initial characterization of fibroblast cluster to sumamrize the clusters to different CAF types.

## CAF markers from literature

```{r}
fibromarkers <- list()

#fibroblast marker in HNSCC from Tirosh paper https://doi.org/10.1016/j.cell.2017.10.044
fibromarkers[["tiroshCAF"]] <- c("FAP","THY1","PDPN","MMP2","MMP11","PDGFRA","PDGFRL","TGFB3","CTGF")
fibromarkers[["tiroshMYF"]] <- c("ACTA2","MCAM","MYLK","MYL9","IL6","PDGFA")
fibromarkers[["tiroshCAFtypes"]] <- c("JUN","FOS","VIM","THY1","FGF7","TGFBR2","TGFBR3","MMP11","CAV1")

#from snapshot CAF paper: http://dx.doi.org/10.1016/j.cell.2020.03.013
fibromarkers[["snapshotCAF"]] <- c("ACTA2","FAP","PDGFRA","PDGFRB","S100A4","MME","C5AR2") #FSP1 is as gene name S100A4, so changed this
fibromarkers[["snapshotcontimmun"]] <- c("ACTA2","FAP") #contractile or immune-modulatory CAs
fibromarkers[["snapshotMYF"]] <- c("TGFB1","TGFB2","IL6") #here included TGFB1 and TGFB2 since TGFB is not a gene name

#fibroblast markers for subtypes from breast cancer paper: https://doi.org/10.1016/j.ccell.2018.01.011
fibromarkers[["BC"]] <- c("ITGB1","FAP","S100A4","ACTA2","PDGFRB","CAV1")

endomarkers <- c("VWF","FABP4")

fm <- c(unique(unlist(fibromarkers)),endomarkers)

#plot markers as UMAP
fibromarkers.p <- lapply(fm,function(x) try(FeaturePlot(D19200.ef,
                                                        features = x,
                                                        reduction = "umap_SCT", 
                                                        cols = c("grey95","blue"), 
                                                        order = T)))
names(fibromarkers.p) <- fm

lapply(fm,function(x) try(ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-",x,".svg"),fibromarkers.p[[x]], width=2.5, height=2.5)))
lapply(fm,function(x) try(ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-",x,"-noleg.svg"),fibromarkers.p[[x]] + theme(legend.position = "none") + ggtitle(""), width=2.5, height=2.5)))

fibromarkers.p
```

```{r}
D19200.ef <- CellCycleScoring(D19200.ef,
                              s.features = cc.genes$s.genes,
                              g2m.features = cc.genes$g2m.genes, 
                              assay="RNA")

#summarize in Heatmap
colorfunc <- function(n) {
  ifelse(n>9,
         colors <- rainbow(n = n),
         colors <- scales::hue_pal()(n)) 
  colors
}
clustercolors <- colorfunc(length(levels(D19200.ef$seurat_clusters_fibro_SCT_PCA)))
names(clustercolors) <- levels(D19200.ef$seurat_clusters_fibro_SCT_PCA)

bottomanno <- HeatmapAnnotation(df = data.frame(Phase=D19200.ef$Phase),
                                col = list(Phase = c(G1=brewer.pal(3,"Set1")[1],
                                                     G2M=brewer.pal(3,"Set1")[2],
                                                     S=brewer.pal(3,"Set1")[3])),
                                which = "column", annotation_name_side = "left")

# Heatmap
# extract gene expression matrix and subset for differentially expressed genes
expr.RNAdata.fm <- as.data.frame(GetAssayData(D19200.ef, assay = "RNA", slot = "data"))[fm,]
expr.RNAscaledata.fm <- as.data.frame(GetAssayData(D19200.ef, assay = "RNA", slot = "scale.data"))[fm,]

# set breaks for Heatmap colors
# Heatmap color breaks are set from 0 to the mean+2*sd of the absolute non-zero scaled expression onto the maximum value, therefore enables setting five colors in total
colorbreaks.RNAscaledata <- c(-max(abs(expr.RNAscaledata.fm)),
                              -(mean(unlist(abs(expr.RNAscaledata.fm))[unlist(abs(expr.RNAscaledata.fm))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.fm))[unlist(abs(expr.RNAscaledata.fm))!=0])),
                              0,mean(unlist(abs(expr.RNAscaledata.fm))[unlist(abs(expr.RNAscaledata.fm))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.fm))[unlist(abs(expr.RNAscaledata.fm))!=0]),
                              max(abs(expr.RNAscaledata.fm)))

colorbreaks.RNAdata <- c(0,mean(unlist(abs(expr.RNAdata.fm))[unlist(abs(expr.RNAdata.fm))!=0]) + 2*sd(unlist(abs(expr.RNAdata.fm))[unlist(abs(expr.RNAdata.fm))!=0]),
                         max(abs(expr.RNAdata.fm)))


#Heatmap clustered by columns and rows, rows split by differential expresison of certain cluster
fm.RNAdata <- Heatmap(expr.RNAdata.fm,
                      name="data",
                      col=circlize::colorRamp2(breaks=colorbreaks.RNAdata, colors = c("black", "yellow","white")),
                      show_column_names = F,
                      show_row_names = T,
                      column_split = D19200.ef$seurat_clusters_fibro_SCT_PCA,
                      cluster_column_slices = F,
                      cluster_columns = T,
                      clustering_method_columns = "ward.D2",
                      clustering_distance_columns = "euclidean",
                      cluster_rows = T,
                      clustering_method_rows = "ward.D2",
                      clustering_distance_rows = "euclidean",
                      bottom_annotation = bottomanno,
                      #graphic options
                      row_names_gp = gpar(fontsize=11),
                      column_names_gp = gpar(fontsize=11),
                      row_title_gp = gpar(fontsize=11),
                      heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                      border = T,
                      use_raster = T,
                      raster_device = "CairoPNG",
                      raster_quality = 4)
fm.RNAscaledata <- Heatmap(expr.RNAscaledata.fm,
                           name="data",
                           col=circlize::colorRamp2(breaks=colorbreaks.RNAscaledata, colors = c("red","purple", "black", "yellow","white")),
                           show_column_names = F,
                           show_row_names = T,
                           column_split = D19200.ef$seurat_clusters_fibro_SCT_PCA,
                           cluster_column_slices = F,
                           cluster_columns = T,
                           clustering_method_columns = "ward.D2",
                           clustering_distance_columns = "euclidean",
                           cluster_rows = T,
                           clustering_method_rows = "ward.D2",
                           clustering_distance_rows = "euclidean",
                           bottom_annotation = bottomanno,
                           #graphic options
                           row_names_gp = gpar(fontsize=11),
                           column_names_gp = gpar(fontsize=11),
                           row_title_gp = gpar(fontsize=11),
                           heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                           border = T,
                           use_raster = T,
                           raster_device = "CairoPNG",
                           raster_quality = 4)


pdf(paste0(OUTPUT,"figs-D19200/D19200-fibroendo-fibmarkers-RNAdata.pdf"), width=20, height=10)
fm.RNAdata
dev.off()
pdf(paste0(OUTPUT,"figs-D19200/D19200-fibroendo-fibmarkers-RNAscaledata.pdf"), width=20, height=10)
fm.RNAscaledata
dev.off()
```

# Differential expression

As recommended by Seurat FAQ, we should use the RNA assay and NOT the pearson residuals of the SCT assay (inside of "scale.data" slot) https://github.com/satijalab/seurat/discussions/4032.

*Slot to use*
It is default and recommended to use the "data" slot of either RNA assay or SCT assay, since the counts are corrected for the sequencing depth. 
Mostly the log Foldchange is used in downstream analysis. The log Foldchange depends on the used slot! 
It is also to note, that the pct.1 and pct.2 values change if you use the RNA or SCT assay.

*Test to use*
Standard method is to use "wilcox". I use MAST since it is also recommended by Theis et al. (https://doi.org/10.15252/msb.20188746) and performs well in a general single cell DE review https://doi.org/10.1038/nmeth.4612, that was recommended in the Seurat tutorial. The test does not create as inflated p-values as with wilcox test.

*PCT option*
The general review also recommends a min.pct of 25%, but since this only excludes genes from testing, I would rather recommend the default of 0.1 to not miss too many information.

```{r}
# Set the Idents to the clustering or splitting of cells for which you want to perform DGE
Idents(D19200.ef) <- NA
Idents(D19200.ef) <- D19200.ef$seurat_clusters_fibro_SCT_PCA

#Differential expression of every cluster against all other cluster
D19200.ef.marker <- FindAllMarkers(D19200.ef,
                                   assay = "RNA",
                                   slot="data", 
                                   test.use = "MAST", #alternative recommendation: MAST
                                   random.seed=1,
                                   logfc.threshold = 0, #shows the test results of all tested genes
                                   min.pct=0.1, #default; prior filtering of tested genes applied!
                                   latent.vars=NULL) # default; only for regression methods
```

# Volcano Plots 

```{r}
library(ggrepel)
ggvolcanopct <- function(data,logFCthres) {
  logFCthres_low <- logFCthres[1]
  logFCthres_up <- logFCthres[2]
  ggplot(data = data,
         aes(x=avg_log2FC,
             y=pct.1-pct.2,
             col=avg_log2FC>logFCthres_up | avg_log2FC<logFCthres_low)) + 
    geom_point() + 
    geom_text_repel(aes(label=ifelse(avg_log2FC>logFCthres_up | avg_log2FC<logFCthres_low,as.character(gene),"")),
                    max.overlaps = 100) +
    scale_color_manual(values = c("TRUE"="red", "FALSE"="grey")) +
    scale_x_continuous(breaks=seq(-10,10,0.5)) +
    scale_y_continuous(breaks=seq(-1,1,0.1)) +
    coord_cartesian(ylim = c(-1,1), xlim=c(-max(abs(c(min(data$avg_log2FC),max(data$avg_log2FC))))-0.1,max(abs(c(min(data$avg_log2FC),max(data$avg_log2FC))))+0.1)) + #symmetric x-axis and y-axis with short buffer
    geom_hline(yintercept = 0, col="black", size=0.1) +
    geom_vline(xintercept = 0, col="black", size=0.1) +
    theme_jb() +
    theme(legend.position = "none", panel.grid.minor = element_blank())
}
#set thres function for selecting genes to mark in volcano plot
thres_func_plot = function(x) {
  c(sort(D19200.ef.marker[D19200.ef.marker$cluster==x,]$avg_log2FC, decreasing = F)[11], 
    sort(D19200.ef.marker[D19200.ef.marker$cluster==x,]$avg_log2FC, decreasing = T)[11])
}

#create volcano plots with pct difference on y-axis
D19200.ef.marker.volpct <- lapply(levels(D19200.ef.marker$cluster), function(x) ggvolcanopct(D19200.ef.marker[D19200.ef.marker$cluster==x,],thres_func_plot(x)) + ggtitle(x))
names(D19200.ef.marker.volpct) <- levels(D19200.ef.marker$cluster)

#save volcano plots
for (i in names(D19200.ef.marker.volpct)) {
  ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-diffexpr-volcanopct-",i,".svg"),D19200.ef.marker.volpct[[i]])
}

D19200.ef.marker.volpct
```

# CAF Annotation and summarizing

based on Combination of UMAP, CAF markers and differential expression results

```{r}
#MYF = Myofibroblast
D19200.ef$type <- as.character(D19200.ef$seurat_clusters_fibro_SCT_PCA)
D19200.ef$type[D19200.ef$type == "0"] <- "CAF1"
D19200.ef$type[D19200.ef$type == "1"] <- "CAF1"
D19200.ef$type[D19200.ef$type == "2"] <- "CAF2"
D19200.ef$type[D19200.ef$type == "3"] <- "CAF2"
D19200.ef$type[D19200.ef$type == "4"] <- "Endothelial Cells"
D19200.ef$type[D19200.ef$type == "5"] <- "CAF3"
D19200.ef$type[D19200.ef$type == "6"] <- "Transition CAF1/2"
D19200.ef$type[D19200.ef$type == "7"] <- "Myoblast-like"

UMAP.SCT.type <- DimPlot(D19200.ef,
                         reduction = "umap_SCT",
                         pt.size = 1,
                         label=T,
                         group.by = "type") + 
  theme_jb_nogrid()

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-type.svg"),UMAP.SCT.type, width=5, height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-type-noleg.svg"),UMAP.SCT.type + theme(legend.position = "none") + ggtitle(""),width=5,height=5)
UMAP.SCT.type
```

# Investigate whether myoblast-like cells are a special type of fibroblasts

Check expression of common markers for fibroblasts identified by differential expression of fibroblast vs all other cellt types

```{r}
#load results of differential expression for all cells and extract genes for fibroblasts
load(paste0(OUTPUT,"RData/D19200-all-diffexpr-marker.RData"))
diffexpr.fibro <- D19200.filt.marker[D19200.filt.marker$cluster=="Fibroblasts",][order(D19200.filt.marker[D19200.filt.marker$cluster=="Fibroblasts",]$avg_log2FC,decreasing = T),]
fdiff.genes <- diffexpr.fibro[diffexpr.fibro$avg_log2FC>1,]$gene #only take genes with logFC greater than 1

#RNAdata
expr.RNAdata.fdiff <- as.data.frame(GetAssayData(D19200.ef, assay = "RNA", slot = "data"))[fdiff.genes,]
# Heatmap color breaks are set from 0 to the mean+2*sd of the absolute non-zero scaled expression onto the maximum value, therefore enables setting five colors in total
colorbreaks.RNAdata.fdiff <- c(0,mean(unlist(abs(expr.RNAdata.fdiff))[unlist(abs(expr.RNAdata.fdiff))!=0]) + 2*sd(unlist(abs(expr.RNAdata.fdiff))[unlist(abs(expr.RNAdata.fdiff))!=0]),
                               max(abs(expr.RNAdata.fdiff)))

#Same Heatmap, but split columns also by clusters
RNAdata.fdiff.hm.split <- Heatmap(expr.RNAdata.fdiff,
                                  name="data",
                                  col=circlize::colorRamp2(breaks=colorbreaks.RNAdata.fdiff, colors = c("black", "yellow","white")),
                                  show_column_names = F,
                                  show_row_names = T,
                                  cluster_columns = T,
                                  cluster_column_slices = F,
                                  clustering_method_columns = "ward.D2",
                                  clustering_distance_columns = "euclidean",
                                  column_split = D19200.ef$type,
                                  cluster_rows = F,
                                  row_order = fdiff.genes[order(diffexpr.fibro[diffexpr.fibro$avg_log2FC>1,]$avg_log2FC,decreasing = T)],
                                  row_title_rot = 0,
                                  column_title_rot = 90,
                                  bottom_annotation = bottomanno,
                                  #graphic options
                                  row_names_gp = gpar(fontsize=11),
                                  column_names_gp = gpar(fontsize=11),
                                  row_title_gp = gpar(fontsize=11),
                                  heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                                  border = T,
                                  use_raster = T,
                                  height=unit(20,"in"),
                                  width=unit(10,"in"),
                                  raster_device = "CairoPNG",
                                  raster_quality = 4)

pdf(paste0(OUTPUT,"figs-D19200/D19200-fibroendo-fdiffgenes-RNAdata-logFC1-split.pdf"), width=20, height=30)
RNAdata.fdiff.hm.split
dev.off()

#Compare expression signature strength defining a fibroblast
D19200.ef <- AddModuleScore(D19200.ef,features = list(fm=fdiff.genes),assay = "RNA",seed = 42,name = "fmsig")

fmsig.v <- VlnPlot(D19200.ef,"fmsig1",assay = "RNA",slot = "data",group.by = "type")

ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-RNA-VLN-fibrosignature.svg"),fmsig.v,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-RNA-VLN-fibrosignature-noleg.svg"),fmsig.v + theme(legend.position = "none") + ggtitle(""),width=5,height=5)

fmsig.v
```

# Investigate genes related to myogenic differentiation

```{r}
#Expression of regulatory factors
#myogenic regulatory factors (MRFs)
MRFs <- c("MYOG", "MYOD1", "MYF5", "MYF6", "MYOPARR", "MEF2A", "MEF2B", "MEF2C", "MEF2D")

MRFs.p <- lapply(MRFs,function(x) try(FeaturePlot(D19200.ef,features = x,reduction = "umap_SCT", cols = c("grey95","blue"), order = T)))
names(MRFs.p) <- MRFs

lapply(MRFs,function(x) try(ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-",x,".svg"),MRFs.p[[x]], width=2.5, height=2.5)))
lapply(MRFs,function(x) try(ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-SCT-UMAP-",x,"-noleg.svg"),MRFs.p[[x]] + theme(legend.position = "none") + ggtitle(""), width=2.5, height=2.5)))

MRFs.p
```

# Differential expression

As recommended by Seurat FAQ, we should use the RNA assay and NOT the pearson residuals of the SCT assay (inside of "scale.data" slot) https://github.com/satijalab/seurat/discussions/4032.

*Slot to use*
It is default and recommended to use the "data" slot of either RNA assay or SCT assay, since the counts are corrected for the sequencing depth. 
Mostly the log Foldchange is used in downstream analysis. The log Foldchange depends on the used slot! 
It is also to note, that the pct.1 and pct.2 values change if you use the RNA or SCT assay.

*Test to use*
Standard method is to use "wilcox". I use MAST since it is also recommended by Theis et al. (https://doi.org/10.15252/msb.20188746) and performs well in a general single cell DE review https://doi.org/10.1038/nmeth.4612, that was recommended in the Seurat tutorial. The test does not create as inflated p-values as with wilcox test.

*PCT option*
The general review also recommends a min.pct of 25%, but since this only excludes genes from testing, I would rather recommend the default of 0.1 to not miss too many information.

```{r}
# Set the Idents to the clustering or splitting of cells for which you want to perform DGE
Idents(D19200.ef) <- NA
Idents(D19200.ef) <- D19200.ef$type

#Differential expression of every cluster against all other cluster
D19200.ef.typemarker <- FindAllMarkers(D19200.ef,
                                       assay = "RNA",
                                       slot="data", 
                                       test.use = "MAST", #alternative recommendation: MAST
                                       random.seed=1,
                                       logfc.threshold = 0, #shows the test results of all tested genes
                                       min.pct=0.1, #default; prior filtering of tested genes applied!
                                       latent.vars=NULL) # default; only for regression methods

#rename transition cells
D19200.ef.typemarker$cluster <- as.character(D19200.ef.typemarker$cluster)
D19200.ef.typemarker$cluster[D19200.ef.typemarker$cluster=="Transition CAF1/2"] <- "Transition"
D19200.ef.typemarker$cluster <- factor(D19200.ef.typemarker$cluster)
```

# Volcano Plots 

```{r}
#set thres function for selecting genes to mark in volcano plot
thres_func_type_plot = function(x) {
  c(sort(D19200.ef.typemarker[D19200.ef.typemarker$cluster==x,]$avg_log2FC, decreasing = F)[11], 
    sort(D19200.ef.typemarker[D19200.ef.typemarker$cluster==x,]$avg_log2FC, decreasing = T)[11])
}

#create volcano plots with pct difference on y-axis
D19200.ef.typemarker.volpct <- lapply(levels(D19200.ef.typemarker$cluster), function(x) ggvolcanopct(D19200.ef.typemarker[D19200.ef.typemarker$cluster==x,],thres_func_type_plot(x)) + ggtitle(x))
names(D19200.ef.typemarker.volpct) <- levels(D19200.ef.typemarker$cluster)

#save volcano plots
for (i in names(D19200.ef.typemarker.volpct)) {
  ggsave(filename = paste0(OUTPUT,"figs-D19200/D19200-fibroendo-diffexpr-type-volcanopct-",i,".svg"),D19200.ef.typemarker.volpct[[i]])
}

D19200.ef.typemarker.volpct
```

# Heatmap of differentially expressed genes

```{r}
#Set logFC threshold (for positive logFC values)
logfcthres.hm <- 1

#format to gene-wise list of diffexpr results
D19200.ef.typemarker.pergene <- lapply(unique(D19200.ef.typemarker$gene), function(x) D19200.ef.typemarker[D19200.ef.typemarker$gene==x,])
names(D19200.ef.typemarker.pergene) <- unique(D19200.ef.typemarker$gene)

#select only entries in genes with logFC > logFC threshold
positivelogFCgenestype <- lapply(D19200.ef.typemarker.pergene, function(x) x[x$avg_log2FC>logfcthres.hm,])
#discard all empty entries, i.e., genes that are not differentially expressed
positivelogFCgenestype <- positivelogFCgenestype[unlist(lapply(positivelogFCgenestype,nrow))!=0]
#converge multiple entries and clusters
positivelogFCgenestype.final <- lapply(positivelogFCgenestype,function(x) {
  ifelse(nrow(x)>1,
         y <- data.frame(gene=unique(x$gene),
                         cluster=toString(x$cluster),
                         avg_log2FC=toString(x$avg_log2FC),
                         numberofstates = nrow(x)),
         y <- cbind(x[,c("gene","cluster","avg_log2FC")], numberofstates = nrow(x)))
  y
})
positivelogFCgenestype.final <- as.data.frame(data.table::rbindlist(positivelogFCgenestype.final))

#for sorting heatmap by foldchange; if multiple foldchanges for one gene, take always the first on (which is from the first cluster)  
positivelogFCgenestype.final.sort <- positivelogFCgenestype.final
if (max(positivelogFCgenestype.final.sort$numberofstates)>1) {
  positivelogFCgenestype.final.sort[positivelogFCgenestype.final.sort$numberofstates>1,]$avg_log2FC <- unlist(lapply(strsplit(positivelogFCgenestype.final.sort[positivelogFCgenestype.final.sort$numberofstates>1,]$avg_log2FC,split = ", "),function(x) x[1]))
  positivelogFCgenestype.final.sort$avg_log2FC <- as.numeric(positivelogFCgenestype.final.sort$avg_log2FC)
}
```

Heatmap of DGE above logFC threshold

```{r}
# extract gene expression matrix and subset for differentially expressed genes
expr.RNAscaledata.type <- as.data.frame(GetAssayData(D19200.ef, assay = "RNA", slot = "scale.data"))
expr.RNAscaledata.type.alldiffgenes <- expr.RNAscaledata.type[as.character(positivelogFCgenestype.final$gene),]

# Heatmap color breaks are set from 0 to the mean+2*sd of the absolute non-zero scaled expression onto the maximum value, therefore enables setting five colors in total
colorbreaks.RNAscaledata.type <- c(-max(abs(expr.RNAscaledata.type.alldiffgenes)),
                                   -(mean(unlist(abs(expr.RNAscaledata.type.alldiffgenes))[unlist(abs(expr.RNAscaledata.type.alldiffgenes))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.type.alldiffgenes))[unlist(abs(expr.RNAscaledata.type.alldiffgenes))!=0])),
                                   0,
                                   mean(unlist(abs(expr.RNAscaledata.type.alldiffgenes))[unlist(abs(expr.RNAscaledata.type.alldiffgenes))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.type.alldiffgenes))[unlist(abs(expr.RNAscaledata.type.alldiffgenes))!=0]),
                                   max(abs(expr.RNAscaledata.type.alldiffgenes)))

#Heatmap clustered by columns and rows, rows split by differential expression of certain cluster; split columns also by clusters
diffexprhm.clustered.split <- Heatmap(expr.RNAscaledata.type.alldiffgenes,
                                      name="scale.data",
                                      col=circlize::colorRamp2(breaks=colorbreaks.RNAscaledata.type, colors = c("red","purple", "black", "yellow","white")),
                                      show_column_names = F,
                                      show_row_names = F,
                                      cluster_columns = T,
                                      cluster_column_slices = F,
                                      clustering_method_columns = "ward.D2",
                                      clustering_distance_columns = "euclidean",
                                      column_split = D19200.ef$type,
                                      row_split = positivelogFCgenestype.final$cluster,
                                      cluster_rows = F,
                                      cluster_row_slices = F,
                                      row_order = rownames(expr.RNAscaledata.type.alldiffgenes)[order(positivelogFCgenestype.final.sort$avg_log2FC,decreasing = T)],
                                      row_title_rot = 0,
                                      column_title_rot = 90,
                                      bottom_annotation = bottomanno,
                                      #graphic options
                                      row_names_gp = gpar(fontsize=11),
                                      column_names_gp = gpar(fontsize=11),
                                      row_title_gp = gpar(fontsize=11),
                                      heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                                      border = T,
                                      use_raster = T,
                                      raster_device = "CairoPNG",
                                      raster_quality = 4,
                                      height = unit(10,"in"),
                                      width = unit(10,"in"))

#recommend not to plot in R console, since the heatmaps are huge
pdf(paste0(OUTPUT,"figs-D19200/D19200-fibroendo-diffexpr-type-HM-RNAscaledata-logfc",as.character(logfcthres.hm),"-clus-split.pdf"), width=20, height=20)
diffexprhm.clustered.split
dev.off()
```

## Heatmaps with only top differentially expressed genes based on log Foldchange

```{r}
#number of topgenes to be plotted
topgenes = 10

#choose those top genes
D19200.ef.typemarker.percluster <- lapply(unique(D19200.ef.typemarker$cluster), function(x) D19200.ef.typemarker[D19200.ef.typemarker$cluster==x,])
names(D19200.ef.typemarker.percluster) <- unique(D19200.ef.typemarker$cluster)

top.logFC <- list()
highlightgenes <- list()
for (cl in names(D19200.ef.typemarker.percluster)) {
  #top avg FC
  top.logFC[[cl]] <- head(D19200.ef.typemarker.percluster[[cl]][order(D19200.ef.typemarker.percluster[[cl]]$avg_log2FC,decreasing = T),]$gene, n=topgenes)
  #create dataframe
  highlightgenes[[cl]] <- data.frame(genes=top.logFC[[cl]], state=cl)
}
# summarize in one dataframe
highlightgenes <- as.data.frame(data.table::rbindlist(highlightgenes))

#subset expression matrix by highlight genes
expr.RNAscaledata.type.hl <- expr.RNAscaledata.type[as.character(highlightgenes$genes),]
#set colorbreaks as above
colorbreaks.RNAscaledata.type.hl <- c(-max(abs(expr.RNAscaledata.type.hl)),
                                      -(mean(unlist(abs(expr.RNAscaledata.type.hl))[unlist(abs(expr.RNAscaledata.type.hl))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.type.hl))[unlist(abs(expr.RNAscaledata.type.hl))!=0])),
                                      0,
                                      mean(unlist(abs(expr.RNAscaledata.type.hl))[unlist(abs(expr.RNAscaledata.type.hl))!=0]) + 2*sd(unlist(abs(expr.RNAscaledata.type.hl))[unlist(abs(expr.RNAscaledata.type.hl))!=0]),
                                      max(abs(expr.RNAscaledata.type.hl)))

#Heatmap clustered by columns and rows, rows split by differential expression of certain cluster; rows split by differential expression of certain cluster, columns split by seurat clusters
diffexprhm.hl.clustered.split <- Heatmap(expr.RNAscaledata.type.hl,
                                         name="scale.data",
                                         col=circlize::colorRamp2(breaks=colorbreaks.RNAscaledata.type.hl, colors = c("red","purple", "black", "yellow","white")),
                                         show_column_names = F,
                                         show_row_names = T,
                                         cluster_columns = T,
                                         cluster_column_slices = F,
                                         clustering_method_columns = "ward.D2",
                                         clustering_distance_columns = "euclidean",
                                         column_split = D19200.ef$type,
                                         row_split = factor(highlightgenes$state),
                                         cluster_rows = T,
                                         cluster_row_slices = F,
                                         clustering_method_rows =  "ward.D2",
                                         clustering_distance_rows = "euclidean",
                                         row_title_rot = 0,
                                         column_title_rot = 90,
                                         bottom_annotation = bottomanno,
                                         #graphic options
                                         row_names_gp = gpar(fontsize=11),
                                         column_names_gp = gpar(fontsize=11),
                                         row_title_gp = gpar(fontsize=11),
                                         heatmap_legend_param = list(title_gp = gpar(fontsize=11), labels_gp = gpar(fontsize=11)),
                                         border = T,
                                         use_raster = T,
                                         raster_device = "CairoPNG",
                                         raster_quality = 4,
                                         height = unit(8.5,"in"),
                                         width = unit(10,"in"))

pdf(paste0(OUTPUT,"figs-D19200/D19200-fibroendo-diffexpr-type-HM-RNAscaledata-highlightgenes-clus-split.pdf"), width=20, height=20)
diffexprhm.hl.clustered.split
dev.off()
```

# Save 

```{r}
#Single Seurat object
save(file = paste0(OUTPUT,"RData/D19200-fibroendo.RData"),D19200.ef)
save(file = paste0(OUTPUT,"RData/D19200-fibroendo-diffexpr-marker.RData"),D19200.ef.marker,D19200.ef.typemarker)
#SessionInfo
writeLines(capture.output(sessionInfo()), paste0(OUTPUT,"RData/16-D19200-fibroendo-sessionInfo.txt"))
```
