---
title: "Downstream - Subsetting"
author: "Kai"
date: "2th June 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat)
library(ggplot2)

source(paste0("~/tscr/kai/projects/01_SCC_scRNA/results/functions.R"))
OUTPUT = "~/tscr/kai/projects/01_SCC_scRNA/results/samples/cohort/all/rp/"
```

# Load Seurat Object

```{r}
load(paste0(OUTPUT,"RData/SCC.RData"))
```

# Overview and recovery rate

expected cells and used cells
```{r}
seqcells <- data.frame(lib = c("1","2","3","4","5","6a","6b","7a","7b"),
                       usedcells=c(10000,10000,10000,18000,10000,26000,26000,26000,20000))
seqcells$expcells <- seqcells$usedcells/2 # we expect a 50% recovery rate
seqcells$cells <- as.numeric(table(SCC$lib))

seqcells$recovery <- seqcells$cells / seqcells$usedcells
seqcells$recovery_text <- paste0(round(seqcells$recovery*100,0)," %")
seqcells$exprecovery  <- seqcells$cells / seqcells$expcells
seqcells$exprecovery_text <- paste0(round(seqcells$exprecovery*100,0)," %")

library(scales)
cells.p <- ggplot(data=reshape2::melt(seqcells[,c("lib","exprecovery_text","expcells","cells")],id.vars = c("lib","exprecovery_text"))) +
  geom_bar(aes(x=lib,
               y=value,
               fill=variable),
           stat="identity",
           position=position_dodge(width = 0.5)) +
  geom_text(aes(x=lib,
               y=value,
               label=exprecovery_text),
            angle=90) +
  labs(x="library",y="cells",fill="recovery") +
  scale_fill_manual(values = c("lightblue","blue")) +
  scale_y_continuous(breaks = seq(0,20000,1000),labels=comma) +
  theme_jb()

ggsave(filename = paste0(OUTPUT,"figs/SCC-expected-recoveryplot.svg"),cells.p,width=5,height=5)

usedcells.p <- ggplot(data=reshape2::melt(seqcells[,c("lib","recovery_text","usedcells","cells")],id.vars = c("lib","recovery_text"))) +
  geom_bar(aes(x=lib,
               y=value,
               fill=variable),
           stat="identity",
           position=position_dodge(width = 0.5)) +
  geom_text(aes(x=lib,
               y=value,
               label=recovery_text),
            angle=90) +
  labs(x="library",y="cells",fill="recovery") +
  scale_fill_manual(values = c("lightblue","blue")) +
  scale_y_continuous(breaks = seq(0,30000,2000),labels=comma) +
  theme_jb()

ggsave(filename = paste0(OUTPUT,"figs/SCC-recoveryplot.svg"),usedcells.p,width=5,height=5)

cells.p
usedcells.p
```

# remove doublets

```{r}
SCC <- subset(SCC,cells = names(SCC$hash.ID[SCC$sample!="Doublet"]))

DefaultAssay(SCC) <- "RNA"
SCC@assays$SCT <- NULL
for (x in names(SCC@reductions)) {SCC@reductions[[x]] <- NULL}
```

# Normalization

## Normalization of RNA assay

RNA assay has three slots: counts, data, scale.data. In the beginning, all three slots have the raw count matrix, after running the normalization below, data and scale.data columns will be normalized, not beforehand (Important to do before differential expression)

```{r SCC_normalization}
#RNA normalization without regressing, necessary for example for differential gene expression not based on SCT.
DefaultAssay(SCC) <- "RNA"
#log-normalization for slot "data"
SCC <- NormalizeData(SCC, 
                     normalization.method = "LogNormalize", #default
                     scale.factor = 10000, #default, scale factor 10000 is recommendation from seurat tutorial
                     assay = "RNA",
                     margin = 1 # default; normalizes across features
) 

#Find Variable Features
SCC <- FindVariableFeatures(SCC, 
                            assay = "RNA",
                            selection.method = "vst", #default
                            nfeatures = 2000 #default; only 2000 , 3000 for SCT promises since to "restore" some more VariableFeatures
)

#scaling for slot "scale.data"
SCC <- ScaleData(SCC,
                 features = VariableFeatures(SCC,assay = "RNA"),
                 do.scale = T,
                 do.center = T,
                 scale.max = 10,
                 assay = "RNA")
```

```{r}
#from Tirosh et al 2015 cell cycle genes, included in Seurat package the "data" slot is always picked, based on RNA assay
SCC <- CellCycleScoring(SCC,
                        s.features = cc.genes.updated.2019$s.genes,
                        g2m.features = cc.genes.updated.2019$g2m.genes, 
                        assay="RNA")
```

## Normalization with SCTransform

```{r SCC_normalization}
SCC <- SCTransform(SCC, 
                   ncells = ncol(SCC),
                   assay="RNA", #default
                   new.assay.name = "SCT", #default; overwrites old SCT
                   do.correct.umi = T, #default change counts slot to corrected counts in new assay
                   variable.features.n = 3000, #default set variable features
                   vars.to.regress = NULL, #default optional variables to regress out
                   do.scale = F, #default scale pearson residuals in scale.data slot to have unit variance
                   do.center = T, #default center pearson residuals in scale.data to have mean zero, needed for PCA
                   return.only.var.genes = T, #default scale.data.matrices output assay only contains variable genes
                   seed.use = 1448145)

# Save SCT normalized seurat object
save(file = paste0(OUTPUT,"RData/SCCfilt-SCTSeurat.RData"),SCC)
```

# Dimension reduction

## PCA

```{r}
SCC <- RunPCA(SCC, 
              npcs = 50,  #number of PCs to use
              assay = "SCT",
              rev.pca = F, # default, Run cell x gene matrix
              weight.by.var = T, #default, Weight cell embedding by variance of each PC!)
              approx = T, #if TRUE, uses irlba instead of prcomp, which is a single value decomposition to approximate PCs, saving computation time (by not calculating all PCs; important for big datasets) and being similar/same to PCs
              features = VariableFeatures(SCC,assay = "SCT"), #default
              reduction.name = "pca_SCT",
              verbose = F)
```

```{r}
#calculate the % variance explained
PCAvar.SCT <- SCC@reductions$pca_SCT@stdev^2 / sum(matrixStats::rowVars(as.matrix(SCC@assays$SCT@scale.data)[VariableFeatures(SCC,assay = "SCT"),]))

#PCA plot
PCAplot.SCT <- DimPlot(SCC, reduction = "pca_SCT", dims = c(1,2), group.by = "patient") + 
  labs(x=paste0("principal component 1 (",signif(PCAvar.SCT[1]*100,digits = 2),"% of variance)"),
       y=paste0("principal component 2 (",signif(PCAvar.SCT[2]*100,digits = 2),"% of variance)")) +
  theme_jb_nogrid()

#elbow plot
PCAelbowplot.SCT <- ggplot(data=data.frame(var=PCAvar.SCT,
                                           PC=1:length(PCAvar.SCT))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=var)) +
  labs(x="principal component", y="% of variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

# cumulative Variance Plot
PCAsumplot.SCT <- ggplot(data=data.frame(cumvar=cumsum(PCAvar.SCT),PC=1:length(PCAvar.SCT))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=cumvar)) +
  labs(x="principal component", y="% of cumulative variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-PCA.svg"),PCAplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-PCA-noleg.svg"),PCAplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-PCA-elbow.svg"),PCAelbowplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-PCA-elbow-noleg.svg"),PCAelbowplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-PCA-variancesum.svg"),PCAsumplot.SCT, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-PCA-variancesum-noleg.svg"),PCAsumplot.SCT + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)

PCAelbowplot.SCT
PCAsumplot.SCT
PCAplot.SCT
```

## Graph-based clustering

```{r FindClusters}
numberofPCs = 20

#based on PCA
SCC <- FindNeighbors(SCC, 
                     dims = 1:numberofPCs, 
                     reduction = "pca_SCT",
                     assay = "SCT")
SCC <- FindClusters(SCC, 
                    resolution = 0.8,
                    random.seed=100)

SCC$seurat_clusters_all_SCT_PCA <- SCC$seurat_clusters
SCC$seurat_clusters <- NULL 
```

## UMAP

```{r UMAP}
SCC <- RunUMAP(SCC,
               dims = 1:numberofPCs,
               assay = "SCT",
               umap.method = "uwot", # Seurat default
               graph=NULL, #default
               reduction="pca_SCT",
               reduction.name = "umap_SCT"
)
```

```{r}
UMAP.SCT.clusPCA <- DimPlot(SCC, 
                            reduction = "umap_SCT", 
                            pt.size = 1, 
                            label = T, 
                            group.by = "seurat_clusters_all_SCT_PCA") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCT.celltype.ss <- DimPlot(SCC, 
                                reduction = "umap_SCT", 
                                pt.size = 1, 
                                label = T, 
                                group.by = "CellType_ss") +  
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCT.sample <- DimPlot(SCC, 
                           reduction = "umap_SCT", 
                           pt.size = 1, 
                           label = F, 
                           group.by = "sample") +  
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCT.patients <- DimPlot(SCC, 
                             reduction = "umap_SCT", 
                             pt.size = 1, 
                             label = F, 
                             group.by = "patient") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 

UMAP.SCT.hashID <- DimPlot(SCC, 
                           reduction = "umap_SCT", 
                           pt.size = 1, 
                           label = F, 
                           group.by = "hash.ID") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 
UMAP.SCT.hashglobal <- DimPlot(SCC, 
                           reduction = "umap_SCT", 
                           pt.size = 1, 
                           label = F, 
                           group.by = "hash.global") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 

UMAP.SCT.Lok <- DimPlot(SCC, 
                        reduction = "umap_SCT", 
                        pt.size = 1, 
                        label = F, 
                        group.by = "Lok") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 

UMAP.SCT.tech <- DimPlot(SCC, 
                         reduction = "umap_SCT", 
                         pt.size = 1, 
                         label = F, 
                         group.by = "technology") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCT.lib <- DimPlot(SCC, 
                        reduction = "umap_SCT", 
                        pt.size = 1, 
                        label = F, 
                        group.by = "lib") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCT.Phase <- DimPlot(SCC, 
                          reduction = "umap_SCT", 
                          pt.size = 1, 
                          label = F, 
                          group.by = "Phase") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCT.QC <- FeaturePlot(SCC, 
                           features = c("nCount_RNA_log10", "nFeature_RNA_log10", "percent.mito", "n.exp.hkgenes","S.Score","G2M.Score"), 
                           reduction = "umap_SCT", 
                           cols = c("grey95", "blue"),
                           pt.size = 1,
                           order = T,
                           slot = "data")

ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-clusPCA.svg"),UMAP.SCT.clusPCA,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-clusPCA-noleg.svg"),UMAP.SCT.clusPCA + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-CellType-ss.svg"),UMAP.SCT.celltype.ss,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-CellType-ss-noleg.svg"),UMAP.SCT.celltype.ss + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-sample.svg"),UMAP.SCT.sample,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-sample-noleg.svg"),UMAP.SCT.sample + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-patients.svg"),UMAP.SCT.patients,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-patients-noleg.svg"),UMAP.SCT.patients + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-hashID.svg"),UMAP.SCT.hashID,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-hashID-noleg.svg"),UMAP.SCT.hashID + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-hashglobal.svg"),UMAP.SCT.hashglobal,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-hashglobal-noleg.svg"),UMAP.SCT.hashglobal + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-Lok.svg"),UMAP.SCT.Lok,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-Lok-noleg.svg"),UMAP.SCT.Lok + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-technology.svg"),UMAP.SCT.tech,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-technology-noleg.svg"),UMAP.SCT.tech + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-lib.svg"),UMAP.SCT.lib,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-lib-noleg.svg"),UMAP.SCT.lib + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-Phase.svg"),UMAP.SCT.Phase, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-Phase-noleg.svg"),UMAP.SCT.Phase + theme(legend.position = "none"), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCT-UMAP-QC.svg"),UMAP.SCT.QC, width = 5, height = 5)

UMAP.SCT.clusPCA
UMAP.SCT.celltype.ss
UMAP.SCT.sample
UMAP.SCT.patients
UMAP.SCT.hashID
UMAP.SCT.hashglobal
UMAP.SCT.Lok
UMAP.SCT.tech
UMAP.SCT.lib
UMAP.SCT.Phase
UMAP.SCT.QC
```

# Regression of cell cycle and mitochondiral percentage

## Normalization with SCTransform

More information: https://doi.org/10.1186/s13059-019-1874-1

```{r SCC_normalization}
SCC <- SCTransform(SCC, 
                   ncells = ncol(SCC),
                   assay="RNA", #default
                   new.assay.name = "SCTccmito", #default; overwrites old SCT
                   do.correct.umi = T, #default change counts slot to corrected counts in new assay
                   variable.features.n = 3000, #default set variable features
                   vars.to.regress = c("S.Score","G2M.Score","percent.mito"), #default optional variables to regress out
                   do.scale = F, #default scale pearson residuals in scale.data slot to have unit variance
                   do.center = T, #default center pearson residuals in scale.data to have mean zero, needed for PCA
                   return.only.var.genes = T, #default scale.data.matrices output assay only contains variable genes
                   seed.use = 1448145)

# Save SCT normalized seurat object
save(file = paste0(OUTPUT,"RData/SCCfilt-SCTSeuratregress.RData"),SCC)
```

# Dimension reduction

## PCA

```{r}
SCC <- RunPCA(SCC, 
              npcs = 50,  #number of PCs to use
              assay = "SCTccmito",
              rev.pca = F, # default, Run cell x gene matrix
              weight.by.var = T, #default, Weight cell embedding by variance of each PC!)
              approx = T, #if TRUE, uses irlba instead of prcomp, which is a single value decomposition to approximate PCs, saving computation time (by not calculating all PCs; important for big datasets) and being similar/same to PCs
              features = VariableFeatures(SCC,assay = "SCTccmito"), #default
              reduction.name = "pca_SCTccmito",
              verbose = F)
```

```{r}
#calculate the % variance explained
PCAvar.SCTccmito <- SCC@reductions$pca_SCTccmito@stdev^2 / sum(matrixStats::rowVars(as.matrix(SCC@assays$SCTccmito@scale.data)[VariableFeatures(SCC,assay = "SCTccmito"),]))

#PCA plot
PCAplot.SCTccmito <- DimPlot(SCC, reduction = "pca_SCTccmito", dims = c(1,2), group.by = "patient") + 
  labs(x=paste0("principal component 1 (",signif(PCAvar.SCTccmito[1]*100,digits = 2),"% of variance)"),
       y=paste0("principal component 2 (",signif(PCAvar.SCTccmito[2]*100,digits = 2),"% of variance)")) +
  theme_jb_nogrid()

#elbow plot
PCAelbowplot.SCTccmito <- ggplot(data=data.frame(var=PCAvar.SCTccmito,
                                                 PC=1:length(PCAvar.SCTccmito))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=var)) +
  labs(x="principal component", y="% of variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

# cumulative variance Plot
PCAsumplot.SCTccmito <- ggplot(data=data.frame(cumvar=cumsum(PCAvar.SCTccmito),PC=1:length(PCAvar.SCTccmito))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=cumvar)) +
  labs(x="principal component", y="% of cumulative variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-PCA.svg"),PCAplot.SCTccmito, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-PCA-noleg.svg"),PCAplot.SCTccmito + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-PCA-elbow.svg"),PCAelbowplot.SCTccmito, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-PCA-elbow-noleg.svg"),PCAelbowplot.SCTccmito + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-PCA-variancesum.svg"),PCAsumplot.SCTccmito, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-PCA-variancesum-noleg.svg"),PCAsumplot.SCTccmito + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)

PCAelbowplot.SCTccmito
PCAsumplot.SCTccmito
PCAplot.SCTccmito
```

# Harmony batch correction

Since the default sigma value gives us mixed myeloid cells and B-Cells, we try to screen this for more accurate clustering catching all cell types.

```{r}
library(harmony)
for (sv in seq(0.1,0.3,0.05)) {
  set.seed(42)
  SCC <- RunHarmony(SCC,
                    reduction="pca_SCTccmito",
                    group.by.vars = "lib",
                    reduction.save = paste0("harmony_SCTccmito_sigma",sv),
                    plot_convergence = T,
                    max.iter.harmony = 30,
                    sigma = sv,
                    assay.use = "SCTccmito",
                    project.dim = F)
}
```

## UMAP

```{r UMAP}
numberofPCs = 20
UMAPharmony.SCTccmito.celltype.ss <- list()
for (sv in seq(0.1,0.3,0.05)) {
  SCC <- RunUMAP(SCC,
                 dims = 1:numberofPCs,
                 assay = "SCTccmito",
                 umap.method = "uwot", # Seurat default
                 graph=NULL, #default
                 reduction=paste0("harmony_SCTccmito_sigma",sv),
                 reduction.name = paste0("humap_SCTccmito_sigma",sv),
                 reduction.key = paste0("humap_SCTccmito_sigma",sv)
  )
  
  UMAPharmony.SCTccmito.celltype.ss[[as.character(sv)]] <- DimPlot(SCC, 
                                                                   reduction = paste0("humap_SCTccmito_sigma",sv), 
                                                                   pt.size = 1, 
                                                                   label = T, 
                                                                   group.by = "CellType_ss") +  
    labs(title = paste0("sigma",sv),x="UMAP_1",y="UMAP_2") +
    theme_jb_nogrid()
  
  ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-sigma",sv,"-CellType-ss-noleg.svg"),UMAPharmony.SCTccmito.celltype.ss[[as.character(sv)]] + theme(legend.position = "none"),width=5,height=5)
}

UMAPharmony.SCTccmito.celltype.ss
```

Choose one sigma: 0.25

```{r}
#based on harmony
DefaultAssay(SCC) <- "SCTccmito"
SCC <- FindNeighbors(SCC,
                     dims = 1:numberofPCs,
                     reduction = "harmony_SCTccmito_sigma0.25",
                     assay = "SCTccmito")
SCC <- FindClusters(SCC,
                    resolution = 0.8,
                    random.seed=100)

SCC$seurat_clusters_all_SCTccmito_PCA_harmony <- SCC$seurat_clusters
SCC$seurat_clusters <- NULL

UMAPharmony.SCTccmito.clusPCA <- DimPlot(SCC, 
                                         reduction = "humap_SCTccmito_sigma0.25", 
                                         pt.size = 1, 
                                         label = T, 
                                         group.by = "seurat_clusters_all_SCTccmito_PCA_harmony") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAPharmony.SCTccmito.sample <- DimPlot(SCC, 
                                        reduction = "humap_SCTccmito_sigma0.25", 
                                        pt.size = 1, 
                                        label = F, 
                                        group.by = "sample") +  
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAPharmony.SCTccmito.patients <- DimPlot(SCC, 
                                          reduction = "humap_SCTccmito_sigma0.25", 
                                          pt.size = 1, 
                                          label = F, 
                                          group.by = "patient") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 

UMAPharmony.SCTccmito.hashID <- DimPlot(SCC, 
                                        reduction = "humap_SCTccmito_sigma0.25", 
                                        pt.size = 1, 
                                        label = F, 
                                        group.by = "hash.ID") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 
UMAPharmony.SCTccmito.hashglobal <- DimPlot(SCC, 
                                            reduction = "humap_SCTccmito_sigma0.25", 
                                            pt.size = 1, 
                                            label = F, 
                                            group.by = "hash.global") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 

UMAPharmony.SCTccmito.Lok <- DimPlot(SCC, 
                                     reduction = "humap_SCTccmito_sigma0.25", 
                                     pt.size = 1, 
                                     label = F, 
                                     group.by = "Lok") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid() 

UMAPharmony.SCTccmito.tech <- DimPlot(SCC, 
                                      reduction = "humap_SCTccmito_sigma0.25", 
                                      pt.size = 1, 
                                      label = F, 
                                      group.by = "technology") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAPharmony.SCTccmito.lib <- DimPlot(SCC, 
                                     reduction = "humap_SCTccmito_sigma0.25", 
                                     pt.size = 1, 
                                     label = F, 
                                     group.by = "lib") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAPharmony.SCTccmito.Phase <- DimPlot(SCC, 
                                       reduction = "humap_SCTccmito_sigma0.25", 
                                       pt.size = 1, 
                                       label = F, 
                                       group.by = "Phase") + 
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()


UMAPharmony.SCTccmito.QC <- FeaturePlot(SCC, 
                                        features = c("nCount_RNA_log10", "nFeature_RNA_log10", "percent.mito", "n.exp.hkgenes","S.Score","G2M.Score"), 
                                        reduction = "humap_SCTccmito_sigma0.25", 
                                        cols = c("grey95", "blue"),
                                        pt.size = 1,
                                        order = T,
                                        slot = "data")

ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-clusPCA.svg"),UMAPharmony.SCTccmito.clusPCA,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-clusPCA-noleg.svg"),UMAPharmony.SCTccmito.clusPCA + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-sample.svg"),UMAPharmony.SCTccmito.sample,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-sample-noleg.svg"),UMAPharmony.SCTccmito.sample + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-patients.svg"),UMAPharmony.SCTccmito.patients,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-patients-noleg.svg"),UMAPharmony.SCTccmito.patients + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-hashID.svg"),UMAPharmony.SCTccmito.hashID,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-hashID-noleg.svg"),UMAPharmony.SCTccmito.hashID + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-hashglobal.svg"),UMAPharmony.SCTccmito.hashglobal,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-hashglobal-noleg.svg"),UMAPharmony.SCTccmito.hashglobal + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-Lok.svg"),UMAPharmony.SCTccmito.Lok,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-Lok-noleg.svg"),UMAPharmony.SCTccmito.Lok + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-technology.svg"),UMAPharmony.SCTccmito.tech,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-technology-noleg.svg"),UMAPharmony.SCTccmito.tech + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-lib.svg"),UMAPharmony.SCTccmito.lib,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-lib-noleg.svg"),UMAPharmony.SCTccmito.lib + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-Phase.svg"),UMAPharmony.SCTccmito.Phase,width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-Phase-noleg.svg"),UMAPharmony.SCTccmito.Phase + theme(legend.position = "none"),width=5,height=5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-QC.svg"),UMAPharmony.SCTccmito.QC, width=10, height=15)

UMAPharmony.SCTccmito.clusPCA
UMAPharmony.SCTccmito.sample
UMAPharmony.SCTccmito.patients
UMAPharmony.SCTccmito.hashID
UMAPharmony.SCTccmito.hashglobal
UMAPharmony.SCTccmito.Lok
UMAPharmony.SCTccmito.tech
UMAPharmony.SCTccmito.lib
UMAPharmony.SCTccmito.Phase
UMAPharmony.SCTccmito.QC
```

# Celltype Annotation

## Marker Gene Expression

```{r}
UMAPharmony.SCTccmito.marker <- lapply(names(celltypemarkers), function(x) FeaturePlot_markers_comb(SCC,celltypemarkers[[x]],reduction = "humap_SCTccmito_sigma0.25"))
names(UMAPharmony.SCTccmito.marker) <- names(celltypemarkers)

lapply(names(UMAPharmony.SCTccmito.marker), function(x) ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-marker-",x,".png"),UMAPharmony.SCTccmito.marker[[x]],dpi=300, width=ifelse(length(celltypemarkers[[x]])>1,10,5), height=5*ceiling(length(celltypemarkers[[x]])/2)))

UMAPharmony.SCTccmito.marker
```

## Assign Cell types based on UMAP harmony clusters

```{r}
Idents(SCC) <- SCC$seurat_clusters_all_SCTccmito_PCA_harmony

SCC <- RenameIdents(SCC,
                    "0" = "T-Cells",
                    "1" = "B-Cells",
                    "2" = "T-Cells",
                    "3" = "T-Cells",
                    "4" = "Tumor",
                    "5" = "T-Cells",
                    "6" = "T-Cells",
                    "7" = "T-Cells",
                    "8" = "Fibroblasts",
                    "9" = "T-Cells",
                    "10" = "Myeloid cells",
                    "11" = "B-Cells",
                    "12" = "T-Cells",
                    "13" = "B-Cells",
                    "14" = "T-Cells",
                    "15" = "T-Cells",
                    "16" = "Plasmablasts",
                    "17" = "B-Cells",
                    "18" = "B-Cells",
                    "19" = "Myeloid cells",
                    "20" = "T-Cells",
                    "21" = "ECs",
                    "22" = "B-Cells",
                    "23" = "pDCs",
                    "24" = "Fibroblasts",
                    "25" = "RBCs",
                    "26" = "B-Cells",
                    "27" = "T-Cells"
                    )

SCC$CellType_SCTccmito_harmony <- Idents(SCC)

UMAPharmony.SCTccmito.celltype <- DimPlot(SCC, 
                                          reduction = "humap_SCTccmito_sigma0.25", 
                                          pt.size = 1, 
                                          label = T, 
                                          group.by = "CellType_SCTccmito_harmony") + 
  theme_jb_nogrid()
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-CellTypeSCTccmitoharmony.svg"),UMAPharmony.SCTccmito.celltype, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/SCCfilt-all-SCTccmito-UMAPharmony-CellTypeSCTccmitoharmony-noleg.svg"),UMAPharmony.SCTccmito.celltype + theme(legend.position = "none"), width = 5, height = 5)
UMAPharmony.SCTccmito.celltype
```

## Save 

```{r}
#Single Seurat object
save(file = paste0(OUTPUT,"RData/SCCfilt.RData"),SCC)
#SessionInfo
writeLines(capture.output(sessionInfo()), paste0(OUTPUT,"RData/SCCfilt-sessionInfo.txt"))
```
