---
title: "patient1 - preprocessing"
author: "Kai"
date: "27th July 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages}
library(Seurat)
library(ggplot2)

#load functions
source(paste0("~/tscr/kai/projects/01_SCC_scRNA/results/functions.R"))
#set output
OUTPUT = "~/tscr/kai/projects/01_SCC_scRNA/results/samples/cohort/patient1/"
```

# Load files and convert to Seurat

```{r}
#Folder of cellranger output
data_dir = "~/tscr/Allgemeine Daten/10x Genomics/data/14405_firstSCC/cellrangerouts_hg38/outs/"
#raw data containing gene expression
patient1.data <- Read10X(paste0(data_dir,"raw_feature_bc_matrix/"))
#Setup Seurat Object for gene Expression
patient1 <- CreateSeuratObject(patient1.data, project = "patient1")
#remove raw data input
rm(patient1.data)
```

# Quality control and Filtering

## Assign the percentage of mitochondrial gene expression and number of expressed housekeeper genes

houskeeping genes are e.g. ribosomal proteins, expressed in nearly every cell and less sensitive to dropout.

```{r}
# relative expression of mitochondrial gene as metric for damaged cells
patient1[["percent.mito"]] <- PercentageFeatureSet(patient1, pattern = "^MT-") 

#read in names of housekeeper genes from Tirosh paper
hkgenes <- read.table("~/tscr/kai/projects/01_SCC_scRNA/db/tirosh_house_keeping.txt", skip = 2)
hkgenes <- as.vector(hkgenes$V1) 
# remove hkgenes that were not found (maybe not in the reference or got lost)
hkgenes.found <- which(toupper(rownames(patient1)) %in% toupper(hkgenes))
# sum of hk genes for each cell
hk_per_cell <- Matrix::colSums(GetAssayData(patient1)[hkgenes.found, ] > 0)
patient1 <- AddMetaData(object = patient1, metadata = hk_per_cell, col.name = "n.exp.hkgenes")
rm(hk_per_cell,hkgenes.found)
```

## Quality Control

We create a central dataframe with all quality metrices for every cell for easier plotting

```{r calculate sequence summary}
barcode_metrics <- data.frame(barcodes=names(patient1$nCount_RNA),
                              nCount =  patient1$nCount_RNA,
                              nFeature = patient1$nFeature_RNA,
                              percent.mito = patient1$percent.mito,
                              n.exp.hkgenes = patient1$n.exp.hkgenes)
```

Than we have four metrices to identify barcodes containing live and usable cells:
* UMI counts (nCount)
* Gene counts (nFeature)
* Mitochondrial gene expression: % of UMIs in mitochondrial genes relative to all genes
* Housekeeper genes: Number of housekeeping genes (97 in total) that have at least 1 UMI count in a cell (no percentage)

It is not helpful to look at all metrics independently since they correlating a lot with each other. So first, the correlations between the metrices are investigated.

### Filter quality metric dataframe for definitely empty barcodes to get readable histograms

First, in this helping dataframe we filter out cells that are definitely empty barcodes since they have only few UMI Counts. These are normally around 70-90% of all barcodes and will give us more informative histograms in downstream processing (Otherwise there will always be a very high peak at around zero).
To examine distribution over all barcodes, we do a cumulative plot showing the % of cells with less than a certain amount of Counts

```{r cumplot}
cumplotdata <- data.frame(x=1:1000, y= unlist(lapply(1:1000,function(x) sum(patient1$nCount_RNA < x) / length(patient1$nCount_RNA < x))))

cumplot.log10 <- ggplot() + 
  geom_point(data=cumplotdata,
             aes(x=log10(x),y=y)) +
  ggtitle("% of barcodes that are lower than x counts") +
  scale_y_continuous(breaks=seq(0,1,0.1)) +
  theme_jb()

cumplot.log10
```

Based on the previous plot we do the actual filtering for around 90% of the cells that are definitely empty in the helping dataframe
```{r}
nCount_lowest <- 5
barcode_metrics <- barcode_metrics[barcode_metrics$nCount > nCount_lowest,]
```

### Examine: UMI counts vs Gene counts

UMI and gene counts are highly correlated with each other. It is not possible to have more genes than counts (since every gene need at least one UMI count).

```{r UMIvsGENE}
UMIvsGENE.mito <- PlotUMIvsGENE(barcode_metrics = barcode_metrics,
                                col = barcode_metrics$percent.mito,
                                colname = "percent.mito")
UMIvsGENE.hk <- PlotUMIvsGENE(barcode_metrics = barcode_metrics,
                              col = barcode_metrics$n.exp.hkgenes,
                              colname = "n.exp.hkgenes")

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsGENE-mito.svg"),UMIvsGENE.mito)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsGENE-hk.svg"),UMIvsGENE.hk)
```

### Examine: mitochondrial gene expression and housekeeper genes vs Counts

```{r mitandhkvsumi}
UMIvsMITO.hk <- PlotUMIvsMITO(barcode_metrics = barcode_metrics,
                              col = barcode_metrics$n.exp.hkgenes,
                              colname = "n.exp.hkgenes")
UMIvsHK.mito <- PlotUMIvsHK(barcode_metrics = barcode_metrics,
                            col = barcode_metrics$percent.mito,
                            colname = "percent.mito")

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsMITO-hk.svg"),UMIvsMITO.hk)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsHK-mito.svg"),UMIvsHK.mito)
```

## Filtering Counts and Features

In general we want to examine all quality metrices and have a special look at certain populations of barcodes in there and want to assign thresholds in that way that these populations are still in there. Good practice here is to have a very permissive filtering first and re-evaluate the filtering after downstream-processing to be more strict.

Rules of Thumb for Filtering:

* for first filtering, be as permissive as possible to not exclude a certain cell population
* Do *not* decide your filtering strategy by looking only on one quality metrics alone - these metrices depend a lot from each other, use the above plots to examine what cut-offs are reasonable. Further downstream, Filtering will be evaluated more in detail.
* First, we only filter for Counts and Features, than repeat QC and set the thresholds for mitochondrial/housekeeper gene expression. This enables better peak identification especially in mitochondrial gene expression, so that more permissive filtering of mitochondrial gene expression combined with housekeeper gene expression is possible.

```{r}
#Cut-offs
#Here you put in your respective cut-off. If you do not want to use a certain cut-off, comment out the respective line in the subsetting command AND the barcode_metrics assignment and set the value to NA
nCount_low <- 500
nCount_high <- NA
nFeature_low <- NA
nFeature_high <- NA

#Subset Seurat object
patient1.filt <- subset(patient1,
                        nCount_RNA > nCount_low
                        #& nCount_RNA < nCount_high
                        #& nFeature_RNA > nFeature_low 
                        #& nFeature_RNA < nFeature_high
)

#assign which barcodes were kept in filtering
barcode_metrics$countfiltering <- NA
barcode_metrics[barcode_metrics$nCount > nCount_low
                #barcode_metrics$nCount < nCount_high & 
                #barcode_metrics$nFeature > nFeature_low &
                #barcode_metrics$nFeature < nFeature_high &
                ,]$countfiltering <- "cell" 
barcode_metrics[!(barcode_metrics$nCount > nCount_low
                  #barcode_metrics$nCount < nCount_high & 
                  #barcode_metrics$nFeature > nFeature_low &
                  #barcode_metrics$nFeature < nFeature_high &
),]$countfiltering <- "background"
```

### Examine: mitochondrial gene expression and housekeeper genes vs Counts

```{r mitandhkvsumi-cf}
UMIvsMITO.cf.hk <- PlotUMIvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$countfiltering=="cell",],
                                 col = barcode_metrics[barcode_metrics$countfiltering=="cell",]$n.exp.hkgenes,
                                 colname = "n.exp.hkgenes")
UMIvsHK.cf.mito <- PlotUMIvsHK(barcode_metrics = barcode_metrics[barcode_metrics$countfiltering=="cell",],
                               col = barcode_metrics[barcode_metrics$countfiltering=="cell",]$percent.mito,
                               colname = "percent.mito")

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsMITO-countfilt-hk.svg"),UMIvsMITO.cf.hk)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsHK-countfilt-mito.svg"),UMIvsHK.cf.mito)
```

### Examine: mitochondrial gene expression vs housekeeper genes

Usually, barcodes with high mitochondrial gene expression and low housekeeper gene expression are regarded as dying cells or cells undergoing large oxidative stress.

```{r mitvshk-cf}
HKvsMITO.cf.UMI <- PlotHKvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$countfiltering=="cell",],
                                col = log10(barcode_metrics[barcode_metrics$countfiltering=="cell",]$nCount),
                                colname = "nCount_log10")

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-HKvsMITO-countfilt-UMI.svg"),HKvsMITO.cf.UMI)
```

## Filtering with mitochondrial and housekeeper genes

```{r}
percent.mito_high <- 10
n.exp.hkgenes_low <- NA

#make sure, you also enable the nCount and Feature filter here!
patient1.filt <- subset(patient1,
                        nCount_RNA > nCount_low
                        #& nCount_RNA < nCount_high
                        #& nFeature_RNA > nFeature_low 
                        #& nFeature_RNA < nFeature_high
                        & percent.mito < percent.mito_high 
                        #& n.exp.hkgenes > n.exp.hkgenes_low
)

#assign which barcodes were kept in filtering
barcode_metrics$manualfiltering <- NA
barcode_metrics[barcode_metrics$nCount > nCount_low &
                  #barcode_metrics$n.exp.hkgenes > n.exp.hkgenes_low & 
                  barcode_metrics$percent.mito < percent.mito_high 
                #barcode_metrics$nCount < nCount_high & 
                #barcode_metrics$nFeature > nFeature_low &
                #barcode_metrics$nFeature < nFeature_high &
                ,]$manualfiltering <- "cell" 
barcode_metrics[!(barcode_metrics$nCount > nCount_low &
                    #barcode_metrics$n.exp.hkgenes > n.exp.hkgenes_low & 
                    barcode_metrics$percent.mito < percent.mito_high
                  #barcode_metrics$nCount < nCount_high & 
                  #barcode_metrics$nFeature > nFeature_low &
                  #barcode_metrics$nFeature < nFeature_high &
),]$manualfiltering <- "background"
```

After filtering for those, we examine what is filtered and whats not filtered.

### Examine: UMI counts vs gene counts after filtering

```{r umivsgene-afterfilt}
#All barcodes, colored by filtered cells with minimum count and feature numbers
UMIvsGENE.filt <- PlotUMIvsGENE(barcode_metrics = barcode_metrics,
                                col = barcode_metrics$manualfiltering,
                                colname = "cells filtered", 
                                xintercept = log10(c(nCount_low,nCount_high)), 
                                yintercept = log10(c(nFeature_low,nFeature_high))) 

#only cell barcodes, colored by percent.mito
UMIvsGENE.cells.mito <- PlotUMIvsGENE(barcode_metrics[barcode_metrics$manualfiltering == "cell",],
                                      col = barcode_metrics[barcode_metrics$manualfiltering == "cell",]$percent.mito,
                                      colname = "percent.mito",
                                      xintercept = log10(c(nCount_low,nCount_high)),
                                      yintercept = log10(c(nFeature_low,nFeature_high)))

# only NOT cells
UMIvsGENE.notcells.mito <-  PlotUMIvsGENE(barcode_metrics[barcode_metrics$manualfiltering != "cell",],
                                          col = barcode_metrics[barcode_metrics$manualfiltering != "cell",]$percent.mito,
                                          colname = "percent.mito",
                                          xintercept = log10(c(nCount_low,nCount_high)),
                                          yintercept = log10(c(nFeature_low,nFeature_high)))

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsGENE-filt.svg"),UMIvsGENE.filt)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsGENE-cells-mito.svg"),UMIvsGENE.cells.mito)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsGENE-notcells-mito.svg"),UMIvsGENE.notcells.mito)
```

### Examine: mitochondrial gene expression and housekeeper genes vs counts after filtering

```{r mitandhkvsumi-afterfilt}
#all barcodes, colored by cells
UMIvsMITO.filt <-  PlotUMIvsMITO(barcode_metrics = barcode_metrics,
                                 col = barcode_metrics$manualfiltering,
                                 colname = "cells filtered",
                                 xintercept = log10(c(nCount_low,nCount_high)),
                                 yintercept = percent.mito_high)
#only cells
UMIvsMITO.cells.hk <-  PlotUMIvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$manualfiltering == "cell",],
                                     col = barcode_metrics[barcode_metrics$manualfiltering == "cell",]$n.exp.hkgenes,
                                     colname = "n.exp.hkgenes",
                                     xintercept = log10(c(nCount_low,nCount_high)),
                                     yintercept = percent.mito_high) 
#only not cells
UMIvsMITO.notcells.hk <-  PlotUMIvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$manualfiltering != "cell",],
                                        col = barcode_metrics[barcode_metrics$manualfiltering != "cell",]$n.exp.hkgenes,
                                        colname = "n.exp.hkgenes",
                                        xintercept = log10(c(nCount_low,nCount_high)),
                                        yintercept = percent.mito_high) 

#all barcodes, colored by cells
UMIvsHK.filt <-  PlotUMIvsHK(barcode_metrics = barcode_metrics,
                             col = barcode_metrics$manualfiltering,
                             colname = "cell",
                             xintercept = log10(c(nCount_low,nCount_high)),
                             yintercept = n.exp.hkgenes_low)
#only cells
UMIvsHK.cells.mito <- PlotUMIvsHK(barcode_metrics = barcode_metrics[barcode_metrics$manualfiltering == "cell",],
                                  col = barcode_metrics[barcode_metrics$manualfiltering == "cell",]$percent.mito,
                                  colname = "percent.mito",
                                  xintercept = log10(c(nCount_low,nCount_high)),
                                  yintercept = n.exp.hkgenes_low) 

#only not cells
UMIvsHK.notcells.mito <- PlotUMIvsHK(barcode_metrics = barcode_metrics[barcode_metrics$manualfiltering != "cell",],
                                     col = barcode_metrics[barcode_metrics$manualfiltering != "cell",]$percent.mito,
                                     colname = "percent.mito",
                                     xintercept = log10(c(nCount_low,nCount_high)),
                                     yintercept = n.exp.hkgenes_low) 

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsMITO-filt.svg"),UMIvsMITO.filt)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsMITO-cells-hk.svg"),UMIvsMITO.cells.hk)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsMITO-notcells-hk.svg"),UMIvsMITO.notcells.hk)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsHK-filt.svg"),UMIvsHK.filt)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsHK-cells-mito.svg"),UMIvsHK.cells.mito)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-UMIvsHK-notcells-mito.svg"),UMIvsHK.notcells.mito)
```

### Examine: percent.mito vs n.exp.hkgenes after filtering

```{r HKvsMITO-afterfilt}
#all barcodes, colored by filtering
HKvsMITO.filt <- PlotHKvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$countfiltering=="cell",],
                              col = barcode_metrics[barcode_metrics$countfiltering=="cell",]$manualfiltering,
                              colname = "cell",
                              xintercept = n.exp.hkgenes_low,
                              yintercept = percent.mito_high)

#only cells, colored by nCounts
HKvsMITO.cells.UMI <- PlotHKvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$manualfiltering == "cell",],
                                   col = log10(barcode_metrics[barcode_metrics$manualfiltering == "cell",]$nCount),
                                   colname = "nCount_log10",
                                   xintercept = n.exp.hkgenes_low, 
                                   yintercept = percent.mito_high)

#only NOT cells, colored by nCounts
HKvsMITO.notcells.UMI <- PlotHKvsMITO(barcode_metrics = barcode_metrics[barcode_metrics$manualfiltering != "cell",],
                                      col = log10(barcode_metrics[barcode_metrics$manualfiltering != "cell",]$nCount),
                                      colname = "nCount_log10",
                                      xintercept = n.exp.hkgenes_low, 
                                      yintercept = percent.mito_high)

ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-HKvsMITO-countfilt-filt.svg"),HKvsMITO.filt)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-HKvsMITO-cells-UMI.svg"),HKvsMITO.cells.UMI)
ggsave(filename = paste0(OUTPUT,"figs/patient1-filtering-HKvsMITO-notcells-UMI.svg"),HKvsMITO.notcells.UMI)
```

# Normalization

The Seurat object is structured in assays, and each assay has three slots containing the actual gene expression matrix:

* slot "counts": raw count matrix
* slot "data": Log-normalized gene expression matrix
* slot "scale.data": scaled data (mean centered and scaled to standard deviation of each feature). Optional, variables can be regressed in this step.

In the beginning, we only have the "RNA" assay and, very important to know, all three slots contain the raw count matrix!
After Normalization of the RNA assay, the slot contain the proper normalized data.

# Normalization of RNA assay

Additionally, sets most variable features.

```{r}
#log-normalization for slot "data"
patient1.filt <- NormalizeData(patient1.filt, 
                               normalization.method = "LogNormalize", #default
                               scale.factor = 10000, #default, scale factor 10000 is recommendation from seurat tutorial
                               assay = "RNA",
                               margin = 1 # default; normalizes across features
) 

#Find Variable Features
patient1.filt <- FindVariableFeatures(patient1.filt, 
                                      assay = "RNA",
                                      selection.method = "vst", #default
                                      nfeatures = 2000 #default; only 2000 , 3000 for SCTccmito promises since to "restore" some more VariableFeatures
)

#scaling for slot "scale.data" by ALL genes, not only most variable ones
patient1.filt <- ScaleData(patient1.filt, 
                           features = VariableFeatures(patient1.filt,assay = "RNA"), 
                           do.scale = T,
                           do.center = T,
                           scale.max = 10, #maximum value for scaling standard deviation, can cause that means are not exactly zero (https://github.com/satijalab/seurat/issues/1166)
                           assay = "RNA")
```

# Determine cell cycle

```{r}
patient1.filt <- CellCycleScoring(patient1.filt,
                                  s.features = cc.genes.updated.2019$s.genes,
                                  g2m.features = cc.genes.updated.2019$g2m.genes, 
                                  assay="RNA")
```

# Normalization using SCTransform

based on the Seurat Tutorial from SCTransform, there can be a benefit of using more PC's in downstream analysis since the normalization regresses out more technical variation. Therefore, also set most variable features for SCTccmito assay.
https://satijalab.org/seurat/v3.1/SCTransform_vignette.html

```{r}
patient1.filt <- SCTransform(patient1.filt, 
                             ncells=ncol(patient1.filt), #default  
                             assay="RNA", #default
                             new.assay.name = "SCTccmito", #default
                             do.correct.umi = T, #default change counts slot to corrected counts in new assay
                             variable.features.n = 3000, #default set variable features
                             vars.to.regress = c("S.Score","G2M.Score","percent.mito"), #default optional variables to regress out
                             do.scale = F, #default scale pearson residuals in scale.data slot to have unit variance
                             do.center = T, #default center pearson residuals in scale.data to have mean zero, needed for PCA
                             return.only.var.genes = T, #default scale.data.matrices output assay only contains variable genes
                             seed.use = 1448145)
```

# Dimensionality reduction and clustering

## Principal Component Analysis (first 50 PCs)

In general, I would recommend to read the general Seurat Tutorial about this:
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#perform-linear-dimensional-reduction-1 

Also, in this Dimensional Reduction Vignette you can find some more information about PCA:
https://satijalab.org/seurat/articles/dim_reduction_vignette.html

PCA Calculation and Plot
```{r PCA, fig.height=6, fig.width=6}
patient1.filt <- RunPCA(patient1.filt,
                        npcs = 50, 
                        assay = "SCTccmito",
                        rev.pca = F, # default, Run cell x gene matrix
                        weight.by.var = T, #default, Weight cell embedding by variance of each PC
                        approx = T, #if TRUE, uses irlba instead of prcomp, which is a single value decomposition to approximate PCs, saving computation time (by not calculating all PCs; important for big datasets) and being similar to PCs. Beware, that approx=F has an error in calculating PCs in optimized Seurat version!
                        features = VariableFeatures(patient1.filt,assay = "SCTccmito"), #default
                        reduction.name = "pca_SCTccmito",
                        verbose = F)
```

Plots

```{r}
#calculate the % variance explained
PCAvar.SCTccmito <- patient1.filt@reductions$pca_SCTccmito@stdev^2 / sum(matrixStats::rowVars(as.matrix(patient1.filt@assays$SCTccmito@scale.data)[VariableFeatures(patient1.filt,assay = "SCTccmito"),]))

#PCA plot
PCAplot.SCTccmito <- DimPlot(patient1.filt, reduction = "pca_SCTccmito", dims = c(1,2), group.by = "orig.ident") + 
  labs(x=paste0("principal component 1 (",signif(PCAvar.SCTccmito[1]*100,digits = 2),"% variance)"),
       y=paste0("principal component 2 (",signif(PCAvar.SCTccmito[2]*100,digits = 2),"% variance)")) +
  theme_jb()

#elbow plot
PCAelbowplot.SCTccmito <- ggplot(data=data.frame(var=PCAvar.SCTccmito,
                                                 PC=1:length(PCAvar.SCTccmito))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=var)) +
  labs(x="principal component", y="% variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

# cumulative Variance Plot
PCAsumplot.SCTccmito <- ggplot(data=data.frame(cumvar=cumsum(PCAvar.SCTccmito),PC=1:length(PCAvar.SCTccmito))[1:50,]) +
  geom_point(mapping = aes(x=PC,y=cumvar)) +
  labs(x="principal component", y="% cumulative variance") +
  scale_x_continuous(breaks=seq(0,50,5)) +
  theme_jb()

ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-PCA.svg"),PCAplot.SCTccmito, width = 5, height = 5)
# without legends
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-PCA-noleg.svg"),PCAplot.SCTccmito + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-PCA-elbow.svg"),PCAelbowplot.SCTccmito, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-PCA-elbow-noleg.svg"),PCAelbowplot.SCTccmito + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-PCA-variancesum.svg"),PCAsumplot.SCTccmito, width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-PCA-variancesum-noleg.svg"),PCAsumplot.SCTccmito + theme(legend.position = "none") + labs(title = NULL),width = 5, height = 5)

PCAplot.SCTccmito
PCAelbowplot.SCTccmito
PCAsumplot.SCTccmito
```

## Graph-based Clustering

We examine based on the Elbowplot how many PC's should be used for clustering and UMAP. Should be a consistent value capturing most variance, so best to choose some PC's after the elbow is flattening.
Recommended default is around 15-30 PC's.

```{r FindClusters}
numberofPCs = 10

patient1.filt <- FindNeighbors(object = patient1.filt,
                               dims = 1:numberofPCs,
                               reduction = "pca_SCTccmito",
                               assay = "SCTccmito")
patient1.filt <- FindClusters(patient1.filt,
                              resolution = 0.8,
                              random.seed = 100)

patient1.filt$seurat_clusters_all_SCTccmito_PCA <- patient1.filt$seurat_clusters
patient1.filt$seurat_clusters <- NULL 
```

## UMAP

```{r UMAP, fig.width=10, fig.height=8}
patient1.filt <- RunUMAP(patient1.filt,
                         dims = 1:numberofPCs,
                         assay = "SCTccmito",
                         umap.method = "uwot", #default
                         graph=NULL, #default
                         reduction="pca_SCTccmito",
                         reduction.name = "umap_SCTccmito")
```

```{r}
patient1.filt$nCount_RNA_log10 <- log10(patient1.filt$nCount_RNA)
patient1.filt$nFeature_RNA_log10 <- log10(patient1.filt$nFeature_RNA)

UMAP.SCTccmito.clusPCA <- DimPlot(patient1.filt, 
                                  reduction = "umap_SCTccmito", 
                                  pt.size = 1, 
                                  label = T, 
                                  group.by = "seurat_clusters_all_SCTccmito_PCA") +
  labs(x="UMAP_1",y="UMAP_2") +
  theme_jb_nogrid()

UMAP.SCTccmito.QC <- FeaturePlot(patient1.filt, 
                                 features = c("nCount_RNA_log10", "nFeature_RNA_log10", "percent.mito", "n.exp.hkgenes","S.Score","G2M.Score"), 
                                 reduction = "umap_SCTccmito", 
                                 cols = c("grey95", "blue"),
                                 pt.size = 1,
                                 order = T,
                                 slot = "data")

ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-UMAP-clusPCA-noleg.svg"),UMAP.SCTccmito.clusPCA + theme(legend.position = "none") + labs(title = NULL), width = 5, height = 5)
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-UMAP-QC.svg"),UMAP.SCTccmito.QC, width=10, height=15)

UMAP.SCTccmito.clusPCA
UMAP.SCTccmito.QC
```

# Cell type identification

## Marker gene expression

```{r}
#defined celltype markers centrally to always use a up-to-date list.
UMAP.SCTccmito.marker <- lapply(names(celltypemarkers), function(x) FeaturePlot_markers_comb(patient1.filt,celltypemarkers[[x]],reduction = "umap_SCTccmito"))
names(UMAP.SCTccmito.marker) <- names(celltypemarkers)

lapply(names(UMAP.SCTccmito.marker), function(x) ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-UMAP-marker-",x,".png"),UMAP.SCTccmito.marker[[x]], width=ifelse(length(celltypemarkers[[x]])>1,10,5), height=5*ceiling(length(celltypemarkers[[x]])/2)))

UMAP.SCTccmito.marker
```

## Celltype Annotation

```{r umap-celltype}
Idents(patient1.filt) <- patient1.filt$seurat_clusters_all_SCTccmito_PCA

patient1.filt <- RenameIdents(patient1.filt,
                              "0" = "Tumor",
                              "1" = "DCs",
                              "2" = "Tumor",
                              "3" = "Fibroblasts",
                              "4" = "Tumor",
                              "5" = "Tumor",
                              "6" = "Macrophages",
                              "7" = "Fibroblasts",
                              "8" = "Fibroblasts",
                              "9" = "Fibroblasts",
                              "10" = "ECs"
)

patient1.filt$CellType <- Idents(patient1.filt)

#new UMAP
UMAP.SCTccmito.celltype <- DimPlot(patient1.filt,
                                   reduction = "umap_SCTccmito", 
                                   pt.size = 1, 
                                   label = T, 
                                   group.by = "CellType") + 
  theme_jb_nogrid()

ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-UMAP-celltype.svg"),UMAP.SCTccmito.celltype, width=5, height=5)
ggsave(filename = paste0(OUTPUT,"figs/patient1-all-SCTccmito-UMAP-celltype-noleg.svg"),UMAP.SCTccmito.celltype + theme(legend.position = "none") + ggtitle(NULL), width=5, height=5)

UMAP.SCTccmito.celltype
```

#Save Output

```{r}
#save Seurat Object
save(file = paste0(OUTPUT,"RData/patient1-filt.RData"),patient1.filt)
#sessioninfo
writeLines(capture.output(sessionInfo()), paste0(OUTPUT,"RData/patient1-all-sessionInfo.txt"))
```
